{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v79)\n",
    "library(ensembldb)\n",
    "suppressMessages(library(dplyr))\n",
    "library('Matrix')\n",
    "set.seed(1)\n",
    "\n",
    "library(RColorBrewer)\n",
    "library(ggplot2)\n",
    "library(pheatmap)\n",
    "library(cowplot)\n",
    "library(reshape2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### set enviroment\n",
    "home = '~/farm/endometrium/data/scATAC/'\n",
    "setwd(home)\n",
    "outdir = paste0(home, 'seurat.output-S2D19/')\n",
    "system(paste0('mkdir -p ', outdir, '/clusters/DARs'))\n",
    "system(paste0('mkdir -p ', outdir, '/clusters/DARs_invivo'))\n",
    "#### \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load data\n",
    "so = readRDS(file = paste0(outdir, '/data/M3_scRNAseq_labelsTransfer.rds'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find differentially accessible peaks between clusters\n",
    "\n",
    "Utilize logistic regression for DA, as suggested by Ntranos et al. 2018 for scRNA-seq data, \n",
    "and add the total number of fragments as a latent variable to mitigate the effect of differential sequencing depth on the result. \n",
    "We can also visualize these chromatin ‘markers’ on a violin plot, feature plot, dot plot, heat map, or any visualization tool in Seurat.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate GRanges object\n",
    "GR = genes(EnsDb.Hsapiens.v79)\n",
    "\n",
    "#switch back to working with peaks instead of gene activities\n",
    "DefaultAssay(so) <- 'peaks'\n",
    "\n",
    "for( cl in rev(levels(Idents(so))) ){\n",
    "    message('Testing ', cl,'\\n')\n",
    "    da_peaks = FindMarkers(so, ident.1 = cl, test.use = \"LR\", \n",
    "                           min.pct = 0.1, \n",
    "                           latent.vars = 'nCount_peaks', \n",
    "                           logfc.threshold = 0.1)\n",
    "    write.csv(da_peaks, file = paste0(outdir, '/clusters/DARs/', cl, '_LR.csv'))\n",
    "    hist(da_peaks$avg_logFC, breaks = 100, main = paste0('cluster ', cl) )\n",
    "    \n",
    "    # Get closest genes\n",
    "#     da_peaks = read.csv(paste0(outdir, '/clusters/DARs/', cl, '_LR.csv'), stringsAsFactors = F)\n",
    "#     open_cl <- da_peaks$X\n",
    "    open_cl <- rownames(da_peaks)\n",
    "    open_cl = gsub('chr', '', open_cl)\n",
    "    \n",
    "    if( length(open_cl) == 0 )\n",
    "        next()\n",
    "    closest_genes_cl <- ClosestFeature(\n",
    "        regions = open_cl,\n",
    "        annotation = GR,\n",
    "        sep = c(':', '-'))\n",
    "    \n",
    "    # add DARs info\n",
    "    idx = match(closest_genes_cl$query_region, open_cl)\n",
    "    closest_genes_cl$Query = open_cl[ idx ]\n",
    "    closest_genes_cl$p_val = da_peaks$p_val[ idx ]\n",
    "    closest_genes_cl$avg_logFC = da_peaks$avg_logFC[ idx ]\n",
    "    closest_genes_cl$pct.1 = da_peaks$pct.1[ idx ]\n",
    "    closest_genes_cl$pct.2 = da_peaks$pct.2[ idx ]\n",
    "    closest_genes_cl$p_val_adj = da_peaks$p_val_adj[ idx ]\n",
    "    \n",
    "    closest_genes_cl = subset(closest_genes_cl, Query == query_region)[, -c(4,6) ]\n",
    "    write.csv(closest_genes_cl, file = paste0(outdir, '/clusters/DARs/', cl, '_closest_genes.csv'), row.names =F)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "so = subset(so, cells = names(Idents(so))[ Idents(so) %in% c('ciliated', 'new_secretory', 'inter_PGRpos') ]  )\n",
    "\n",
    "for( cl in c('new_secretory', 'inter_PGRpos', 'ciliated') ){\n",
    "    message('Testing ', cl,'\\n')\n",
    "    da_peaks = FindMarkers(so, ident.1 = cl, test.use = \"LR\", \n",
    "                           min.pct = 0.1, \n",
    "                           latent.vars = 'nCount_peaks', \n",
    "                           logfc.threshold = 0.1)\n",
    "    write.csv(da_peaks, file = paste0(outdir, '/clusters/DARs_invivo/', cl, '_LR.csv'))\n",
    "    hist(da_peaks$avg_logFC, breaks = 100, main = paste0('cluster ', cl) )\n",
    "    \n",
    "    # Get closest genes\n",
    "#     da_peaks = read.csv(paste0(outdir, '/clusters/DARs/', cl, '_LR.csv'), stringsAsFactors = F)\n",
    "#     open_cl <- da_peaks$X\n",
    "    open_cl <- rownames(da_peaks)\n",
    "    open_cl = gsub('chr', '', open_cl)\n",
    "    \n",
    "    if( length(open_cl) == 0 )\n",
    "        next()\n",
    "    closest_genes_cl <- ClosestFeature(\n",
    "        regions = open_cl,\n",
    "        annotation = GR,\n",
    "        sep = c(':', '-'))\n",
    "    \n",
    "    # add DARs info\n",
    "    idx = match(closest_genes_cl$query_region, open_cl)\n",
    "    closest_genes_cl$Query = open_cl[ idx ]\n",
    "    closest_genes_cl$p_val = da_peaks$p_val[ idx ]\n",
    "    closest_genes_cl$avg_logFC = da_peaks$avg_logFC[ idx ]\n",
    "    closest_genes_cl$pct.1 = da_peaks$pct.1[ idx ]\n",
    "    closest_genes_cl$pct.2 = da_peaks$pct.2[ idx ]\n",
    "    closest_genes_cl$p_val_adj = da_peaks$p_val_adj[ idx ]\n",
    "    \n",
    "    closest_genes_cl = subset(closest_genes_cl, Query == query_region)[, -c(4,6) ]\n",
    "    write.csv(closest_genes_cl, file = paste0(outdir, '/clusters/DARs_invivo/', cl, '_closest_genes.csv'), row.names =F)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create_contingency_table = function(cl, GenSet){\n",
    "#   cl_GenSet = intersect(GenSet, cl) %>% length(.)\n",
    "#   nocl_GenSet =  unique(GenSet) %>% length(.)\n",
    "#   noGS = setdiff(allGS, GenSet)\n",
    "#   cl_outGenSet = intersect(noGS, cl) %>% length(.)\n",
    "#   nocl_outGenSet = unique(noGS) %>% length(.)\n",
    "#   m = matrix(c(cl_GenSet, nocl_GenSet,  cl_outGenSet,  nocl_outGenSet), \n",
    "#              ncol = 2, dimnames = list(c('Genset', 'All'), c('in_cl', 'total')) )\n",
    "#   return(m)  \n",
    "# }\n",
    "\n",
    "create_contingency_table = function(cl, GenSet){\n",
    "  cl_GenSet = intersect(GenSet, cl) %>% length(.)\n",
    "  Allcl_GenSet =  intersect(GenSet, GENES) %>% length(.)\n",
    "  noGenSet = setdiff(allGS, GenSet)\n",
    "  cl_AllGenSet = intersect(noGenSet, cl) %>% length(.)\n",
    "  Allcl_AllGenSet = intersect(noGenSet, GENES) %>% length(.)\n",
    "  m = matrix(c(cl_GenSet,  cl_AllGenSet,  Allcl_GenSet, Allcl_AllGenSet), byrow = T,\n",
    "             ncol = 2, dimnames = list(c('cl', 'Allcl'), c('GenSet', 'noGenSet')) )\n",
    "  return(m)  \n",
    "}\n",
    "\n",
    "\n",
    "enrichment = function(cl, GenSet){\n",
    "  funtable = create_contingency_table(cl, GenSet)\n",
    "  out = NULL\n",
    "  if ( funtable[1,1] > 1 ){\n",
    "    ft = fisher.test(funtable)\n",
    "    pvalue = ft$p.value\n",
    "    estimate = ft$estimate\n",
    "    conf.int = ft$conf.int\n",
    "    out = data.frame(pvalue= pvalue, odds.ratio = estimate, min_confint = conf.int[1], max_confint = conf.int[2], n=funtable[1,1], stringsAsFactors = F)\n",
    "  return(out)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "GENES = list()\n",
    "for( cl in as.character(0:12) ){\n",
    "    closest_genes_cl = read.csv(paste0(outdir, '/clusters/DARs/',cl,'_closest_genes.csv'), stringsAsFactors = F)\n",
    "    closest_genes_cl = subset(closest_genes_cl, gene_name != \"\" )\n",
    "    closest_genes_cl = subset(closest_genes_cl, distance < 5000)\n",
    "    GENES[[cl]] = closest_genes_cl$gene_name\n",
    "}\n",
    "GENES = unlist(GENES) %>% unique(.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viper_gset = get(load('~/farm/gsea/genesets/dorotheav2-top10scoring_VentoLab20200121.rdata'))\n",
    "GS_pos = lapply(viper_gset, function(x) names(x$tfmode)[ as.numeric(x$tfmode) >= 0 ] )\n",
    "GS_neg = lapply(viper_gset, function(x) names(x$tfmode)[ as.numeric(x$tfmode) < 0 ] )\n",
    "GS_pos = lapply(GS_pos, intersect, GENES)\n",
    "GS_neg = lapply(GS_neg, intersect, GENES)\n",
    "GS_pos = GS_pos[ sapply(GS_pos, length) >= 3 ]\n",
    "GS_neg = GS_neg[ sapply(GS_neg, length ) >= 3 ]\n",
    "names(GS_pos) = paste0(names(GS_pos), '_pos')\n",
    "names(GS_neg) = paste0(names(GS_neg), '_neg')\n",
    "GS = append(GS_pos, GS_neg)\n",
    "allGS = unlist(GS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viper_gset = get(load('~/farm/gsea/genesets/dorotheav2-top10scoring_VentoLab20200121.rdata'))\n",
    "# GS = lapply(viper_gset, function(x) names(x$tfmode) )\n",
    "# GS = lapply(GS, intersect, GENES)\n",
    "# GS = GS[ sapply(GS, length) >= 3 ]\n",
    "# allGS = unlist(GS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GS$CSRNP1_E_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = list()\n",
    "\n",
    "for( cl in as.character(0:12) ){\n",
    "    message(cl)\n",
    "    closest_genes_cl = read.csv(paste0(outdir, '/clusters/DARs/',cl,'_closest_genes.csv'), stringsAsFactors = F)\n",
    "    closest_genes_cl = subset(closest_genes_cl, gene_name != \"\" )\n",
    "    closest_genes_cl = subset(closest_genes_cl, distance < 5000)\n",
    "    closest_genes_cl = closest_genes_cl[ order(closest_genes_cl$p_val_adj, decreasing = F), ]\n",
    "    closest_genes_cl = closest_genes_cl[ ! duplicated(closest_genes_cl$gene_name),  ] \n",
    "    closest_genes_cl = subset(closest_genes_cl, p_val_adj < 0.05 & avg_logFC > 0.5)\n",
    "    \n",
    "    cl_genes = closest_genes_cl$gene_name\n",
    "    ENRCH = lapply(GS, enrichment, cl=cl_genes) \n",
    "    print(cl_genes[1:5])\n",
    "    fENRCH = ENRCH[ ! sapply(ENRCH, is.null) ] \n",
    "    df_ENRCH = melt(fENRCH, id.vars = colnames(fENRCH[[1]]) )\n",
    "    names(df_ENRCH)[6] = 'TF'\n",
    "    df_ENRCH$cl = cl\n",
    "    if( nrow(df_ENRCH) > 0 )\n",
    "        results[[cl]] = df_ENRCH\n",
    "}\n",
    "df = melt(results, id.vars = names(results[[1]]))\n",
    "df = df[ order(df$pvalue) , -ncol(df) ]\n",
    "write.csv(df, file = paste0(outdir, '/clusters/DARs/TFs_activities_GSEA.csv'), row.names = F, quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = list()\n",
    "\n",
    "for( cl in c('new_secretory', 'ciliated', 'inter_PGRpos')  ){\n",
    "    message(cl)\n",
    "    closest_genes_cl = read.csv(paste0(outdir, '/clusters/DARs_invivo/',cl,'_closest_genes.csv'), stringsAsFactors = F)\n",
    "    closest_genes_cl = subset(closest_genes_cl, gene_name != \"\" )\n",
    "    closest_genes_cl = subset(closest_genes_cl, distance < 5000)\n",
    "    closest_genes_cl = closest_genes_cl[ order(closest_genes_cl$p_val_adj, decreasing = F), ]\n",
    "    closest_genes_cl = closest_genes_cl[ ! duplicated(closest_genes_cl$gene_name),  ] \n",
    "    closest_genes_cl = subset(closest_genes_cl, p_val_adj < 0.01 & avg_logFC > 0)\n",
    "    \n",
    "    cl_genes = closest_genes_cl$gene_name\n",
    "    ENRCH = lapply(GS, enrichment, cl=cl_genes) \n",
    "    fENRCH = ENRCH[ ! sapply(ENRCH, is.null) ] \n",
    "    print(fENRCH$CSRNP1_E)\n",
    "    df_ENRCH = melt(fENRCH, id.vars = colnames(fENRCH[[1]]) )\n",
    "    names(df_ENRCH)[6] = 'TF'\n",
    "    df_ENRCH$cl = cl\n",
    "    if( nrow(df_ENRCH) > 0 )\n",
    "        results[[cl]] = df_ENRCH\n",
    "}\n",
    "df = melt(results, id.vars = names(results[[1]]))\n",
    "df = df[ order(df$pvalue) , -ncol(df) ]\n",
    "write.csv(df, file = paste0(outdir, '/clusters/DARs_invivo/TFs_activities_GSEA.csv'), row.names = F, quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Genesets\n",
    "viper_gset = get(load('~/farm/gsea/genesets/msigDB_viper.rdata'))\n",
    "# names(viper_gset) = paste0('msigDB_', names(viper_gset))\n",
    "viper_gset = viper_gset[ c(grep(\"KEGG\", names(viper_gset)), grep(\"REACTOME\", names(viper_gset))) ]\n",
    "# viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/harmonizome_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/TFrole_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/GObpUniprot_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/cellUniprot_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/DevelopUniprot_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/tissueExpressionUniprot_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/UniprotKeywords_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/Seuratccgenes_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/bgee_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/ensembl_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/ovarianDevelopment_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/stevant2019_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/zhang2018_oocyte_GC_DEsignatures_viper.rdata')))\n",
    "viper_gset = append(x = viper_gset,  get(load('~/farm/gsea/genesets/TestisDevelopment_viper.rdata')))\n",
    "idx2remove = unlist(sapply(c('DISEASE', 'TUMOR', 'CANCER', 'CARCINOMA', 'OMA_', 'OMA ', 'RESISTANCE', 'NEOPLA', 'VIRAL', 'CARIES','BRAIN', \n",
    "                             'RAPAMYCIN', 'INFLUENZA', 'CARCINOGENESIS', 'ASTHMA', '_FLU', 'SERUM', 'DISORDER', 'PROBIOTIC', '_DAY',\n",
    "                             'ONCO', 'E_COLI', 'INFECTION', 'THERAPY', 'LEUKEMIA', 'TROPHY', 'OID_', 'MOUSE', 'FUSION', 'CIGenSetLATIN', 'METASTASIS',\n",
    "                             'INJURY', 'TRANGenSetLANT', 'BACTER', 'FUNG', 'HEART', 'FUNG', 'TRANGenSetLANT', '_DN$'), \n",
    "                           grep, x= toupper(names(viper_gset))))\n",
    "viper_gset = viper_gset[ - idx2remove ]\n",
    "\n",
    "\n",
    "GS = lapply(viper_gset, function(x) names(x$tfmode) )\n",
    "GS = lapply(GS, intersect, GENES)\n",
    "GS = GS[ sapply(GS, length) > 3 ]\n",
    "allGS = unlist(GS) %>% unique(.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = list()\n",
    "\n",
    "for( cl in rev(levels(Idents(so))) ){\n",
    "    message(cl)\n",
    "    closest_genes_cl = read.csv(paste0(outdir, '/clusters/DARs/',cl,'_closest_genes.csv'), stringsAsFactors = F)\n",
    "    closest_genes_cl = subset(closest_genes_cl, gene_name != \"\" )\n",
    "    closest_genes_cl = subset(closest_genes_cl, distance < 5000)\n",
    "    closest_genes_cl = closest_genes_cl[ order(closest_genes_cl$p_val_adj, decreasing = F), ]\n",
    "    closest_genes_cl = closest_genes_cl[ ! duplicated(closest_genes_cl$gene_name),  ] \n",
    "    closest_genes_cl = subset(closest_genes_cl, p_val_adj < 0.05 & avg_logFC > 0)\n",
    "    \n",
    "    cl_genes = closest_genes_cl$gene_name\n",
    "    ENRCH = lapply(GS, enrichment, cl=cl_genes) \n",
    "    fENRCH = ENRCH[ ! sapply(ENRCH, is.null) ] \n",
    "    df_ENRCH = melt(fENRCH, id.vars = colnames(fENRCH[[1]]) )\n",
    "    names(df_ENRCH)[6] = 'pathway'\n",
    "    df_ENRCH$cl = cl\n",
    "    if( nrow(df_ENRCH) > 0 )\n",
    "        results[[cl]] = df_ENRCH\n",
    "}\n",
    "df = melt(results, id.vars = names(results[[1]]))\n",
    "df = df[ order(df$pvalue) , -ncol(df) ]\n",
    "write.csv(df, file = paste0(outdir, '/clusters/DARs/pathways_GSEA.csv'), row.names = F, quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = list()\n",
    "\n",
    "for( cl in rev(levels(Idents(so))) ){\n",
    "    message(cl)\n",
    "    closest_genes_cl = read.csv(paste0(outdir, '/clusters/DARs_invivo/',cl,'_closest_genes.csv'), stringsAsFactors = F)\n",
    "    closest_genes_cl = subset(closest_genes_cl, gene_name != \"\" )\n",
    "    closest_genes_cl = subset(closest_genes_cl, distance < 5000)\n",
    "    closest_genes_cl = closest_genes_cl[ order(closest_genes_cl$p_val_adj, decreasing = F), ]\n",
    "    closest_genes_cl = closest_genes_cl[ ! duplicated(closest_genes_cl$gene_name),  ] \n",
    "    closest_genes_cl = subset(closest_genes_cl, p_val_adj < 0.05 & avg_logFC > 0)\n",
    "    \n",
    "    cl_genes = closest_genes_cl$gene_name\n",
    "    ENRCH = lapply(GS, enrichment, cl=cl_genes) \n",
    "    fENRCH = ENRCH[ ! sapply(ENRCH, is.null) ] \n",
    "    df_ENRCH = melt(fENRCH, id.vars = colnames(fENRCH[[1]]) )\n",
    "    names(df_ENRCH)[6] = 'pathway'\n",
    "    df_ENRCH$cl = cl\n",
    "    if( nrow(df_ENRCH) > 0 )\n",
    "        results[[cl]] = df_ENRCH\n",
    "}\n",
    "df = melt(results, id.vars = names(results[[1]]))\n",
    "df = df[ order(df$pvalue) , -ncol(df) ]\n",
    "write.csv(df, file = paste0(outdir, '/clusters/DARs_invivo/pathways_GSEA.csv'), row.names = F, quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
