{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'treeio':\n",
      "  method     from\n",
      "  root.phylo ape \n",
      "\n",
      "See system.file(\"LICENSE\", package=\"MotifDb\") for use restrictions.\n",
      "\n",
      "\n",
      "Attaching package: 'Matrix'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:TFBSTools':\n",
      "\n",
      "    Matrix\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "********************************************************\n",
      "\n",
      "Note: As of version 1.0.0, cowplot does not change the\n",
      "\n",
      "  default ggplot2 theme anymore. To recover the previous\n",
      "\n",
      "  behavior, execute:\n",
      "  theme_set(theme_cowplot())\n",
      "\n",
      "********************************************************\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'cowplot'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:patchwork':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(ensembldb)\n",
    "library(JASPAR2018)\n",
    "library(TFBSTools)\n",
    "library(patchwork)\n",
    "library(universalmotif)\n",
    "library(MotifDb)\n",
    "library(TFBSTools)\n",
    "library(patchwork)\n",
    "\n",
    "suppressMessages(library(dplyr))\n",
    "library('Matrix')\n",
    "set.seed(1)\n",
    "\n",
    "library(RColorBrewer)\n",
    "library(ggplot2)\n",
    "library(pheatmap)\n",
    "library(cowplot)\n",
    "library(reshape2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### set enviroment\n",
    "home = '~/farm/endometrium/data/scATAC/'\n",
    "setwd(home)\n",
    "outdir = paste0(home, 'seurat.output-S2D19/')\n",
    "system(paste0('mkdir -p ', outdir, '/clusters'))\n",
    "#### \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load data\n",
    "so = readRDS(file = paste0(outdir, 'data/M3_scRNAseq_labelsTransfer.rds'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct a Motif object and add it to the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(so) <- 'peaks'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of motif position frequency matrices from the JASPAR, HOCOMOCO, HOMER and SwissRegulon databases\n",
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2018,\n",
    "  opts = list(species = 9606, all_versions = FALSE)\n",
    ")\n",
    "# add data from hocomoco A\n",
    "tf_motif_hoco=subset(MotifDb, organism=='Hsapiens' & dataSource=='HOCOMOCOv11-core-A')\n",
    "pfmHoco = convert_motifs(tf_motif_hoco, 'TFBSTools-PWMatrix')\n",
    "pfmHoco_names =  sapply(pfmHoco, function(x) x@ID)\n",
    "names(tf_motif_hoco) = pfmHoco_names\n",
    "for (i in 1:length(pfmHoco) ){\n",
    "    # Convert into a PCM\n",
    "    x = t(pfmHoco[[i]]@profileMatrix) - apply(pfmHoco[[i]]@profileMatrix, 2, min) # normalize per column\n",
    "    pfmHoco[[i]]@profileMatrix = round(t(x / apply(x, 1, max))*100) # mulitply x100 and round\n",
    "    class(pfmHoco[[i]]@profileMatrix) <- 'integer'\n",
    "    pfm[[pfmHoco_names[i]]] <- pfmHoco[[i]]\n",
    "}\n",
    "# add data from hocomoco B\n",
    "tf_motif_hoco=subset(MotifDb, organism=='Hsapiens' & dataSource=='HOCOMOCOv11-core-B')\n",
    "pfmHoco = convert_motifs(tf_motif_hoco, 'TFBSTools-PWMatrix')\n",
    "pfmHoco_names =  sapply(pfmHoco, function(x) x@ID)\n",
    "names(tf_motif_hoco) = pfmHoco_names\n",
    "for (i in 1:length(pfmHoco) ){\n",
    "    # Convert into a PCM\n",
    "    x = t(pfmHoco[[i]]@profileMatrix) - apply(pfmHoco[[i]]@profileMatrix, 2, min) # normalize per column\n",
    "    pfmHoco[[i]]@profileMatrix = round(t(x / apply(x, 1, max))*100) # mulitply x100 and round\n",
    "    class(pfmHoco[[i]]@profileMatrix) <- 'integer'\n",
    "    pfm[[pfmHoco_names[i]]] <- pfmHoco[[i]]\n",
    "}\n",
    "# add data from hocomoco C\n",
    "tf_motif_hoco=subset(MotifDb, organism=='Hsapiens' & dataSource=='HOCOMOCOv11-core-C')\n",
    "pfmHoco = convert_motifs(tf_motif_hoco, 'TFBSTools-PWMatrix')\n",
    "pfmHoco_names =  sapply(pfmHoco, function(x) x@ID)\n",
    "names(tf_motif_hoco) = pfmHoco_names\n",
    "for (i in 1:length(pfmHoco) ){\n",
    "    # Convert into a PCM\n",
    "    x = t(pfmHoco[[i]]@profileMatrix) - apply(pfmHoco[[i]]@profileMatrix, 2, min) # normalize per column\n",
    "    pfmHoco[[i]]@profileMatrix = round(t(x / apply(x, 1, max))*100) # mulitply x100 and round\n",
    "    class(pfmHoco[[i]]@profileMatrix) <- 'integer'\n",
    "    pfm[[pfmHoco_names[i]]] <- pfmHoco[[i]]\n",
    "}\n",
    "# add data from SwissRegulon                       \n",
    "tf_motif_SwissRegulon=subset(MotifDb, organism=='Hsapiens' & dataSource=='SwissRegulon')\n",
    "pfmSwissRegulon = convert_motifs(tf_motif_SwissRegulon, 'TFBSTools-PWMatrix')\n",
    "pfmSwissRegulon_names =  sapply(pfmSwissRegulon, function(x) x@ID)\n",
    "names(tf_motif_SwissRegulon) = pfmSwissRegulon_names\n",
    "for (i in 1:length(pfmSwissRegulon) ){\n",
    "    # Convert into a PCM\n",
    "    x = t(pfmSwissRegulon[[i]]@profileMatrix) - apply(pfmSwissRegulon[[i]]@profileMatrix, 2, min) # normalize per column\n",
    "    pfmSwissRegulon[[i]]@profileMatrix = round(t(x / apply(x, 1, max))*100) # mulitply x100 and round\n",
    "    class(pfmSwissRegulon[[i]]@profileMatrix) <- 'integer'\n",
    "    pfm[[pfmSwissRegulon_names[i]]] <- pfmSwissRegulon[[i]]\n",
    "}\n",
    "# add data from HOMER                       \n",
    "tf_motif_HOMER=subset(MotifDb, dataSource=='HOMER')\n",
    "pfmHOMER = convert_motifs(tf_motif_HOMER, 'TFBSTools-PWMatrix')\n",
    "pfmHOMER_names =  sapply(pfmHOMER, function(x) x@ID)\n",
    "names(pfmHOMER) = paste0( sapply( strsplit(pfmHOMER_names, '\\\\('), head, 1), '.HOMER')\n",
    "pfmHOMER = pfmHOMER[ grep('[a-z]',  names(pfmHOMER), invert = T) ] # remove non-human (lowercase)         \n",
    "for (i in 1:length(pfmHOMER) ){\n",
    "    # Convert into a PCM\n",
    "    x = t(pfmHOMER[[i]]@profileMatrix) - apply(pfmHOMER[[i]]@profileMatrix, 2, min) # normalize per column\n",
    "    pfmHOMER[[i]]@profileMatrix = round(t(x / apply(x, 1, max))*100) # mulitply x100 and round\n",
    "    class(pfmHOMER[[i]]@profileMatrix) <- 'integer'\n",
    "    pfmHOMER[[i]]@name = gsub('.HOMER', '', names(pfmHOMER)[i])\n",
    "    pfm[[names(pfmHOMER)[i]]] <- pfmHOMER[[i]]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add name in id\n",
    "idx =  grep('^MA', names(pfm) )\n",
    "for (i in idx ){\n",
    "    pfm[[i]]@ID = paste0(pfm[[i]]@name, '.', pfm[[i]]@ID)\n",
    "    names(pfm)[i] = pfm[[i]]@ID\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove pcm with NAs\n",
    "pfm = pfm[ ! sapply(pfm, function(x) any(is.na(x@profileMatrix))) ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1681"
      ],
      "text/latex": [
       "1681"
      ],
      "text/markdown": [
       "1681"
      ],
      "text/plain": [
       "[1] 1681"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "HOCOMOCOv11-core-A HOCOMOCOv11-core-B HOCOMOCOv11-core-C              HOMER \n",
       "               181                 84                135                146 \n",
       "      SwissRegulon \n",
       "               683 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(pfm)\n",
    "table(unlist(sapply(pfm, function(x) x@tags$dataSource)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scan the DNA sequence of each peak for the presence of each motif\n",
    "motif.matrix <- CreateMotifMatrix(\n",
    "  features = StringToGRanges(rownames(so), sep = c(\":\", \"-\")),\n",
    "  pwm = pfm,\n",
    "  genome = 'hg38',\n",
    "  sep = c(\":\", \"-\")\n",
    ")\n",
    "# Create a new Mofif object to store the results\n",
    "motif <- CreateMotifObject(\n",
    "  data = motif.matrix,\n",
    "  pwm = pfm\n",
    ")\n",
    "# Add the Motif object to the assay\n",
    "so[['peaks']] <- AddMotifObject(\n",
    "  object = so[['peaks']],\n",
    "  motif.object = motif\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute  sequence characteristics of the peaks: GC content, sequence length, and dinucleotide frequency\n",
    "so <- RegionStats(\n",
    "  object = so,\n",
    "  genome = BSgenome.Hsapiens.UCSC.hg38,\n",
    "  sep = c(\":\", \"-\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding active TFmotifs - ChromVAR\n",
    "\n",
    "This tends to give similar results as performing an enrichment test on differentially accessible peaks between the cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- RunChromVAR(\n",
    "  object = so,\n",
    "  genome = BSgenome.Hsapiens.UCSC.hg38\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(so,  paste0(outdir, '/data/M4_TFmotif_activities.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# so = readRDS(paste0(outdir, '/data/M4_TFmotif_activities.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif2names = lapply(pfm, function(x) x@name)\n",
    "names(motif2names) = sapply(pfm, function(x) x@ID) %>% gsub('_', '-', .)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(so) <- 'chromvar'\n",
    "for( cl in levels(Idents(so)) ){\n",
    "  message('Testing ', cl,'\\n')\n",
    "  differential.activity = FindMarkers(so, ident.1 = cl, test.use = \"LR\", only.pos = TRUE, latent.vars = 'nCount_peaks', \n",
    "                                      logfc.threshold = 0.1, min.pct = 0.1)\n",
    "  differential.activity$TF = motif2names[ rownames(differential.activity) ] %>% as.character(.)\n",
    "  write.csv(differential.activity, file = paste0(outdir, '/clusters/', cl, '_differential.activity_LR.csv'))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare Ciliated and new secretory vs PGR+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "retrieve_top10 = function(differential.activity, n = 10, name = F){\n",
    "    # Fix homer names\n",
    "  idx = grep('HOMER', differential.activity$X)\n",
    "  differential.activity$TF[ idx ] = gsub('.HOMER', '', differential.activity$X[ idx ])\n",
    "  # Fix hocomoco synonyms\n",
    "  differential.activity$TF2 = differential.activity$TF\n",
    "  idx = differential.activity$TF %in% syn2gen$V1\n",
    "  differential.activity$TF2[ idx ] =  syn2gen$V2[ match(differential.activity$TF[ idx ], syn2gen$V1) ] \n",
    "  differential.activity = differential.activity[ ! duplicated(differential.activity$TF2) , ]\n",
    "  differential.activity = subset(differential.activity, avg_logFC > 0.5)\n",
    "  differential.activity = subset(differential.activity, p_val_adj < 0.01)\n",
    "    \n",
    "  tfs = head(differential.activity$X, n = n) \n",
    "  if(name)\n",
    "    tfs = head(differential.activity$TF2, n = n) \n",
    "  return(tfs)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load synonyms to annorate Tfs to their symbol\n",
    "syn2gen = read.delim('~/secondary2primary_genename.annot', stringsAsFactors = F, header = F)\n",
    "syn2gen = syn2gen[ order(syn2gen$V1 == syn2gen$V2, decreasing = T) , ]\n",
    "syn2gen = syn2gen[ ! duplicated(syn2gen$V1) , ] # remove secondaries with several synonyms ( different from the primary )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(so) <- 'chromvar'\n",
    "TOP10 = list()\n",
    "All = list()\n",
    "\n",
    "# PGR\n",
    "differential.activity = FindMarkers(so, \n",
    "                                    ident.1 = c('inter_PGRpos'), ident.2 = c('new_secretory', 'ciliated'), \n",
    "                                    test.use = \"LR\", only.pos = TRUE, latent.vars = 'nCount_peaks', \n",
    "                                    logfc.threshold = 0.1, min.pct = 0.1)\n",
    "differential.activity$TF = motif2names[ rownames(differential.activity) ] %>% as.character(.)\n",
    "differential.activity$X = rownames(differential.activity)\n",
    "write.csv(differential.activity, file = paste0(outdir, '/clusters/PGR_vs_other_differential.activity_LR.csv'))\n",
    "TOP10[['PGR']] = retrieve_top10(differential.activity)\n",
    "All[['PGR']] = retrieve_top10(differential.activity, n = 20000, name = T)\n",
    "\n",
    "# ciliated\n",
    "differential.activity = FindMarkers(so, \n",
    "                                    ident.1 = 'ciliated', ident.2 = c('inter_PGRpos', 'new_secretory'), \n",
    "                                    test.use = \"LR\", only.pos = TRUE, latent.vars = 'nCount_peaks', \n",
    "                                    logfc.threshold = 0.1, min.pct = 0.1)\n",
    "differential.activity$TF = motif2names[ rownames(differential.activity) ] %>% as.character(.)\n",
    "differential.activity$X = rownames(differential.activity)\n",
    "write.csv(differential.activity, file = paste0(outdir, '/clusters/ciliated_vs_other_differential.activity_LR.csv'))\n",
    "TOP10[['ciliated']] = retrieve_top10(differential.activity)\n",
    "All[['ciliated']] = retrieve_top10(differential.activity, n = 20000, name = T)\n",
    "\n",
    "# new_secretory\n",
    "differential.activity = FindMarkers(so, \n",
    "                                    ident.1 = 'new_secretory', ident.2 = c('inter_PGRpos', 'ciliated'), \n",
    "                                    test.use = \"LR\", only.pos = TRUE, latent.vars = 'nCount_peaks', \n",
    "                                    logfc.threshold = 0.1, min.pct = 0.1)\n",
    "differential.activity$TF = motif2names[ rownames(differential.activity) ] %>% as.character(.)\n",
    "differential.activity$X = rownames(differential.activity)\n",
    "write.csv(differential.activity, file = paste0(outdir, '/clusters/newsecretory_vs_other_differential.activity_LR.csv'))\n",
    "TOP10[['new_secretory']] = retrieve_top10(differential.activity)\n",
    "All[['new_secretory']] = retrieve_top10(differential.activity, n = 20000, name = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "soF = subset(so, cells = names(Idents(so))[ Idents(so) %in% c('inter_PGRpos', 'ciliated') ] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(soF) = factor(Idents(soF), levels =  rev(c('inter_PGRpos', 'ciliated')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DoHeatmap(soF, features = unlist(TOP10), slot = 'data') + scale_fill_gradientn(colors = c(\"grey\", \"red\"), )\n",
    "# ggsave(device = 'pdf', file = paste0(outdir, '/clusters/heatmap_TFmotifs_invivo.pdf'), width = 12, height = 10)\n",
    "DotPlot(object = soF, features = rev(unlist(TOP10))) + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))\n",
    "ggsave(device = 'pdf', file = paste0(outdir, '/clusters/dotplot_TFmotifs_invivo.pdf'), width = 12, height = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TFmotifs = c('E2F4-HUMAN.H11MO.0.A', \"E2F4.MA0470.1\", \"E2F4.HOMER\", 'PGR.HOMER','ESR1.MA0112.3', \n",
    "            'TP73.SwissRegulon', 'FOXJ1.SwissRegulon', \"RFX2-HUMAN.H11MO.0.A\", 'RFX2.MA0600.2', \n",
    "            'HNF1B-HUMAN.H11MO.0.A', 'CEBPB-HUMAN.H11MO.0.A', \"CEBPB.MA0466.2\", \"CEBPB.SwissRegulon\")\n",
    "# DoHeatmap(soF, features = TFmotifs, slot = 'data') + scale_fill_gradientn(colors = c(\"white\", \"red\"), )\n",
    "# ggsave(device = 'pdf', file = paste0(outdir, '/clusters/heatmap_TFmotifs_invivo_intersectRNA.pdf'), width = 12, height = 10)\n",
    "DotPlot(object = soF, features = rev(TFmotifs)) + \n",
    "# theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + \n",
    "theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))\n",
    "ggsave(device = 'pdf', file = paste0(outdir, '/clusters/dotplot_TFmotifs_invivo_intersectRNA.pdf'), width = 7, height = 5.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot top TF motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing 11\n",
      "\n",
      "\n",
      "Testing 10\n",
      "\n",
      "\n",
      "Testing 9\n",
      "\n",
      "\n",
      "Testing 8\n",
      "\n",
      "\n",
      "Testing 7\n",
      "\n",
      "\n",
      "Testing 6\n",
      "\n",
      "\n",
      "Testing 5\n",
      "\n",
      "\n",
      "Testing 4\n",
      "\n",
      "\n",
      "Testing 3\n",
      "\n",
      "\n",
      "Testing 2\n",
      "\n",
      "\n",
      "Testing 1\n",
      "\n",
      "\n",
      "Testing 0\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`11`</dt>\n",
       "\t\t<dd></dd>\n",
       "\t<dt>$`10`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'FOXA2'</li><li>'DBP'</li><li>'RFX'</li><li>'RFX1'</li><li>'FOXJ2'</li><li>'RFX7'</li><li>'SNAI2'</li><li>'TCF12'</li><li>'TFE2'</li><li>'FOXH1'</li><li>'REST'</li><li>'ONECUT2'</li><li>'NHLH1'</li><li>'BRCA1'</li><li>'BHA15'</li><li>'MESP1'</li><li>'ESRRA'</li><li>'MYF6'</li><li>'MSC'</li><li>'OLIG2'</li><li>'GLI1'</li><li>'ZIC1'</li><li>'ZKSC1'</li><li>'MYB'</li><li>'REST-NRSF'</li><li>'MTF1'</li><li>'ZIC4'</li><li>'ZSCAN4'</li><li>'ZIC3'</li><li>'TBX21'</li><li>'ESRRB'</li><li>'NFIA'</li><li>'CTCFL'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`9`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'FOXJ1'</li><li>'PRDM6'</li><li>'IRF1'</li><li>'DMRT1'</li><li>'ZN770'</li><li>'IRF9'</li><li>'FOXJ3'</li><li>'FOXA1:AR'</li><li>'STAT1::STAT2'</li><li>'FUBP1'</li><li>'LEF1'</li><li>'BPTF'</li><li>'SMCA5'</li><li>'ZN121'</li><li>'IRF3'</li><li>'IKZF2'</li><li>'HNF4G'</li><li>'IRF7'</li><li>'NFATC2'</li><li>'HNF4A'</li><li>'NFATC3'</li><li>'TCF3'</li><li>'RBPJ'</li><li>'NFKB1'</li><li>'NR2F1'</li><li>'RXRG'</li><li>'RXRB'</li><li>'COUP-TFII'</li><li>'TF65'</li><li>'NR1H2::RXRA'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`8`</dt>\n",
       "\t\t<dd></dd>\n",
       "\t<dt>$`7`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ZNF740'</li><li>'ZNF281'</li><li>'HOXB5'</li><li>'ZN281'</li><li>'ZNF263'</li><li>'RXRA'</li><li>'EWSR1-FLI1'</li><li>'ZBT17'</li><li>'SP4'</li><li>'E2F7'</li><li>'KLF15'</li><li>'PPARG'</li><li>'E2F1'</li><li>'WRNIP1'</li><li>'ZSC22'</li><li>'E2F6'</li><li>'PTF1A'</li><li>'NKX25'</li><li>'ETV5'</li><li>'ZN263'</li><li>'GTF2I'</li><li>'ZN467'</li><li>'MAZ'</li><li>'VEZF1'</li><li>'SPIB'</li><li>'ZN341'</li><li>'RARA'</li><li>'PATZ1'</li><li>'ETS2'</li><li>'E2F3'</li><li>'SUH'</li><li>'TBX3'</li><li>'INSM1'</li><li>'NFIC'</li><li>'MTA3'</li><li>'KLF12'</li><li>'ZN257'</li><li>'LYL1'</li><li>'ZN816'</li><li>'TFAP2A'</li><li>'ZN320'</li><li>'AP2A'</li><li>'ZN436'</li><li>'ZNF467'</li><li>'SRBP2'</li><li>'THAP1'</li><li>'GATA2'</li><li>'SNAI1'</li><li>'NR1I3'</li><li>'ZN563'</li><li>'SMAD1'</li><li>'TLX1'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`6`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ELF5'</li><li>'ELF3'</li><li>'EHF'</li><li>'GRHL2'</li><li>'ELF1'</li><li>'KLF5'</li><li>'KLF4'</li><li>'ELF4'</li><li>'ETV6'</li><li>'ERG'</li><li>'SPDEF'</li><li>'KLF1'</li><li>'ELF2'</li><li>'ETV2'</li><li>'ELK4'</li><li>'GABPA'</li><li>'ETV7'</li><li>'KLF9'</li><li>'ETS1'</li><li>'NF-E2'</li><li>'BACH2'</li><li>'GRHL1'</li><li>'SPIC'</li><li>'ZBTB7A'</li><li>'PBX3'</li><li>'JUN'</li><li>'FOSL1::JUND(var.2)'</li><li>'ATF4'</li><li>'TFCP2'</li><li>'CEBPA'</li><li>'CEBPB'</li><li>'IRF8'</li><li>'CEBPE'</li><li>'NFIL3'</li><li>'IRF4'</li><li>'DDIT3'</li><li>'CEBP:AP1'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`5`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CTCF'</li><li>'NFYB'</li><li>'RAD21'</li><li>'SMC3'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`4`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TEAD2'</li><li>'TEAD4'</li><li>'TEAD1'</li><li>'ISL1'</li><li>'TEAD'</li><li>'TEAD3'</li><li>'HOXB3'</li><li>'ZFHX3'</li><li>'ESX1'</li><li>'HOXD3'</li><li>'PBX2'</li><li>'HESX1'</li><li>'GBX1'</li><li>'HXB7'</li><li>'HOXB7'</li><li>'HXA9'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`3`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'BACH1'</li><li>'PAX6'</li><li>'PAX9'</li><li>'RELA'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`2`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TP53'</li><li>'TP63'</li><li>'TP73'</li><li>'BCL11A'</li><li>'RUNX3'</li><li>'TFCP2L1'</li><li>'AHR'</li><li>'ARNT2'</li><li>'EGR3'</li><li>'MAF::NFE2'</li><li>'DUXA'</li><li>'TCF7L2'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`1`</dt>\n",
       "\t\t<dd></dd>\n",
       "\t<dt>$`0`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SPI1'</li><li>'HOXA4'</li><li>'PITX1'</li><li>'PITX2'</li><li>'PAX5'</li><li>'POU2F3'</li><li>'HOXA13'</li><li>'ONECUT3'</li><li>'HNF1A'</li><li>'ZNF384'</li><li>'RCOR1'</li><li>'TFAP2D'</li><li>'NFE2'</li><li>'FOSB'</li><li>'BATF::JUN'</li><li>'SP8'</li><li>'SMARCC1'</li><li>'FOS::JUND'</li><li>'FOSL1::JUN'</li><li>'FOS::JUN'</li><li>'FOSL2::JUN'</li><li>'FOSL1::JUND'</li><li>'FOSB::JUNB'</li><li>'FOSL2::JUND'</li><li>'JUN::JUNB'</li><li>'FOSL1::JUNB'</li><li>'FOSL2::JUNB'</li><li>'FOS::JUNB'</li><li>'E2F4'</li><li>'MBD2'</li><li>'NRF1'</li><li>'HIC1'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`11`] \n",
       "\\item[\\$`10`] \\begin{enumerate*}\n",
       "\\item 'FOXA2'\n",
       "\\item 'DBP'\n",
       "\\item 'RFX'\n",
       "\\item 'RFX1'\n",
       "\\item 'FOXJ2'\n",
       "\\item 'RFX7'\n",
       "\\item 'SNAI2'\n",
       "\\item 'TCF12'\n",
       "\\item 'TFE2'\n",
       "\\item 'FOXH1'\n",
       "\\item 'REST'\n",
       "\\item 'ONECUT2'\n",
       "\\item 'NHLH1'\n",
       "\\item 'BRCA1'\n",
       "\\item 'BHA15'\n",
       "\\item 'MESP1'\n",
       "\\item 'ESRRA'\n",
       "\\item 'MYF6'\n",
       "\\item 'MSC'\n",
       "\\item 'OLIG2'\n",
       "\\item 'GLI1'\n",
       "\\item 'ZIC1'\n",
       "\\item 'ZKSC1'\n",
       "\\item 'MYB'\n",
       "\\item 'REST-NRSF'\n",
       "\\item 'MTF1'\n",
       "\\item 'ZIC4'\n",
       "\\item 'ZSCAN4'\n",
       "\\item 'ZIC3'\n",
       "\\item 'TBX21'\n",
       "\\item 'ESRRB'\n",
       "\\item 'NFIA'\n",
       "\\item 'CTCFL'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`9`] \\begin{enumerate*}\n",
       "\\item 'FOXJ1'\n",
       "\\item 'PRDM6'\n",
       "\\item 'IRF1'\n",
       "\\item 'DMRT1'\n",
       "\\item 'ZN770'\n",
       "\\item 'IRF9'\n",
       "\\item 'FOXJ3'\n",
       "\\item 'FOXA1:AR'\n",
       "\\item 'STAT1::STAT2'\n",
       "\\item 'FUBP1'\n",
       "\\item 'LEF1'\n",
       "\\item 'BPTF'\n",
       "\\item 'SMCA5'\n",
       "\\item 'ZN121'\n",
       "\\item 'IRF3'\n",
       "\\item 'IKZF2'\n",
       "\\item 'HNF4G'\n",
       "\\item 'IRF7'\n",
       "\\item 'NFATC2'\n",
       "\\item 'HNF4A'\n",
       "\\item 'NFATC3'\n",
       "\\item 'TCF3'\n",
       "\\item 'RBPJ'\n",
       "\\item 'NFKB1'\n",
       "\\item 'NR2F1'\n",
       "\\item 'RXRG'\n",
       "\\item 'RXRB'\n",
       "\\item 'COUP-TFII'\n",
       "\\item 'TF65'\n",
       "\\item 'NR1H2::RXRA'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`8`] \n",
       "\\item[\\$`7`] \\begin{enumerate*}\n",
       "\\item 'ZNF740'\n",
       "\\item 'ZNF281'\n",
       "\\item 'HOXB5'\n",
       "\\item 'ZN281'\n",
       "\\item 'ZNF263'\n",
       "\\item 'RXRA'\n",
       "\\item 'EWSR1-FLI1'\n",
       "\\item 'ZBT17'\n",
       "\\item 'SP4'\n",
       "\\item 'E2F7'\n",
       "\\item 'KLF15'\n",
       "\\item 'PPARG'\n",
       "\\item 'E2F1'\n",
       "\\item 'WRNIP1'\n",
       "\\item 'ZSC22'\n",
       "\\item 'E2F6'\n",
       "\\item 'PTF1A'\n",
       "\\item 'NKX25'\n",
       "\\item 'ETV5'\n",
       "\\item 'ZN263'\n",
       "\\item 'GTF2I'\n",
       "\\item 'ZN467'\n",
       "\\item 'MAZ'\n",
       "\\item 'VEZF1'\n",
       "\\item 'SPIB'\n",
       "\\item 'ZN341'\n",
       "\\item 'RARA'\n",
       "\\item 'PATZ1'\n",
       "\\item 'ETS2'\n",
       "\\item 'E2F3'\n",
       "\\item 'SUH'\n",
       "\\item 'TBX3'\n",
       "\\item 'INSM1'\n",
       "\\item 'NFIC'\n",
       "\\item 'MTA3'\n",
       "\\item 'KLF12'\n",
       "\\item 'ZN257'\n",
       "\\item 'LYL1'\n",
       "\\item 'ZN816'\n",
       "\\item 'TFAP2A'\n",
       "\\item 'ZN320'\n",
       "\\item 'AP2A'\n",
       "\\item 'ZN436'\n",
       "\\item 'ZNF467'\n",
       "\\item 'SRBP2'\n",
       "\\item 'THAP1'\n",
       "\\item 'GATA2'\n",
       "\\item 'SNAI1'\n",
       "\\item 'NR1I3'\n",
       "\\item 'ZN563'\n",
       "\\item 'SMAD1'\n",
       "\\item 'TLX1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`6`] \\begin{enumerate*}\n",
       "\\item 'ELF5'\n",
       "\\item 'ELF3'\n",
       "\\item 'EHF'\n",
       "\\item 'GRHL2'\n",
       "\\item 'ELF1'\n",
       "\\item 'KLF5'\n",
       "\\item 'KLF4'\n",
       "\\item 'ELF4'\n",
       "\\item 'ETV6'\n",
       "\\item 'ERG'\n",
       "\\item 'SPDEF'\n",
       "\\item 'KLF1'\n",
       "\\item 'ELF2'\n",
       "\\item 'ETV2'\n",
       "\\item 'ELK4'\n",
       "\\item 'GABPA'\n",
       "\\item 'ETV7'\n",
       "\\item 'KLF9'\n",
       "\\item 'ETS1'\n",
       "\\item 'NF-E2'\n",
       "\\item 'BACH2'\n",
       "\\item 'GRHL1'\n",
       "\\item 'SPIC'\n",
       "\\item 'ZBTB7A'\n",
       "\\item 'PBX3'\n",
       "\\item 'JUN'\n",
       "\\item 'FOSL1::JUND(var.2)'\n",
       "\\item 'ATF4'\n",
       "\\item 'TFCP2'\n",
       "\\item 'CEBPA'\n",
       "\\item 'CEBPB'\n",
       "\\item 'IRF8'\n",
       "\\item 'CEBPE'\n",
       "\\item 'NFIL3'\n",
       "\\item 'IRF4'\n",
       "\\item 'DDIT3'\n",
       "\\item 'CEBP:AP1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`5`] \\begin{enumerate*}\n",
       "\\item 'CTCF'\n",
       "\\item 'NFYB'\n",
       "\\item 'RAD21'\n",
       "\\item 'SMC3'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`4`] \\begin{enumerate*}\n",
       "\\item 'TEAD2'\n",
       "\\item 'TEAD4'\n",
       "\\item 'TEAD1'\n",
       "\\item 'ISL1'\n",
       "\\item 'TEAD'\n",
       "\\item 'TEAD3'\n",
       "\\item 'HOXB3'\n",
       "\\item 'ZFHX3'\n",
       "\\item 'ESX1'\n",
       "\\item 'HOXD3'\n",
       "\\item 'PBX2'\n",
       "\\item 'HESX1'\n",
       "\\item 'GBX1'\n",
       "\\item 'HXB7'\n",
       "\\item 'HOXB7'\n",
       "\\item 'HXA9'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`3`] \\begin{enumerate*}\n",
       "\\item 'BACH1'\n",
       "\\item 'PAX6'\n",
       "\\item 'PAX9'\n",
       "\\item 'RELA'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`2`] \\begin{enumerate*}\n",
       "\\item 'TP53'\n",
       "\\item 'TP63'\n",
       "\\item 'TP73'\n",
       "\\item 'BCL11A'\n",
       "\\item 'RUNX3'\n",
       "\\item 'TFCP2L1'\n",
       "\\item 'AHR'\n",
       "\\item 'ARNT2'\n",
       "\\item 'EGR3'\n",
       "\\item 'MAF::NFE2'\n",
       "\\item 'DUXA'\n",
       "\\item 'TCF7L2'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`1`] \n",
       "\\item[\\$`0`] \\begin{enumerate*}\n",
       "\\item 'SPI1'\n",
       "\\item 'HOXA4'\n",
       "\\item 'PITX1'\n",
       "\\item 'PITX2'\n",
       "\\item 'PAX5'\n",
       "\\item 'POU2F3'\n",
       "\\item 'HOXA13'\n",
       "\\item 'ONECUT3'\n",
       "\\item 'HNF1A'\n",
       "\\item 'ZNF384'\n",
       "\\item 'RCOR1'\n",
       "\\item 'TFAP2D'\n",
       "\\item 'NFE2'\n",
       "\\item 'FOSB'\n",
       "\\item 'BATF::JUN'\n",
       "\\item 'SP8'\n",
       "\\item 'SMARCC1'\n",
       "\\item 'FOS::JUND'\n",
       "\\item 'FOSL1::JUN'\n",
       "\\item 'FOS::JUN'\n",
       "\\item 'FOSL2::JUN'\n",
       "\\item 'FOSL1::JUND'\n",
       "\\item 'FOSB::JUNB'\n",
       "\\item 'FOSL2::JUND'\n",
       "\\item 'JUN::JUNB'\n",
       "\\item 'FOSL1::JUNB'\n",
       "\\item 'FOSL2::JUNB'\n",
       "\\item 'FOS::JUNB'\n",
       "\\item 'E2F4'\n",
       "\\item 'MBD2'\n",
       "\\item 'NRF1'\n",
       "\\item 'HIC1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`11`\n",
       ":   \n",
       "$`10`\n",
       ":   1. 'FOXA2'\n",
       "2. 'DBP'\n",
       "3. 'RFX'\n",
       "4. 'RFX1'\n",
       "5. 'FOXJ2'\n",
       "6. 'RFX7'\n",
       "7. 'SNAI2'\n",
       "8. 'TCF12'\n",
       "9. 'TFE2'\n",
       "10. 'FOXH1'\n",
       "11. 'REST'\n",
       "12. 'ONECUT2'\n",
       "13. 'NHLH1'\n",
       "14. 'BRCA1'\n",
       "15. 'BHA15'\n",
       "16. 'MESP1'\n",
       "17. 'ESRRA'\n",
       "18. 'MYF6'\n",
       "19. 'MSC'\n",
       "20. 'OLIG2'\n",
       "21. 'GLI1'\n",
       "22. 'ZIC1'\n",
       "23. 'ZKSC1'\n",
       "24. 'MYB'\n",
       "25. 'REST-NRSF'\n",
       "26. 'MTF1'\n",
       "27. 'ZIC4'\n",
       "28. 'ZSCAN4'\n",
       "29. 'ZIC3'\n",
       "30. 'TBX21'\n",
       "31. 'ESRRB'\n",
       "32. 'NFIA'\n",
       "33. 'CTCFL'\n",
       "\n",
       "\n",
       "\n",
       "$`9`\n",
       ":   1. 'FOXJ1'\n",
       "2. 'PRDM6'\n",
       "3. 'IRF1'\n",
       "4. 'DMRT1'\n",
       "5. 'ZN770'\n",
       "6. 'IRF9'\n",
       "7. 'FOXJ3'\n",
       "8. 'FOXA1:AR'\n",
       "9. 'STAT1::STAT2'\n",
       "10. 'FUBP1'\n",
       "11. 'LEF1'\n",
       "12. 'BPTF'\n",
       "13. 'SMCA5'\n",
       "14. 'ZN121'\n",
       "15. 'IRF3'\n",
       "16. 'IKZF2'\n",
       "17. 'HNF4G'\n",
       "18. 'IRF7'\n",
       "19. 'NFATC2'\n",
       "20. 'HNF4A'\n",
       "21. 'NFATC3'\n",
       "22. 'TCF3'\n",
       "23. 'RBPJ'\n",
       "24. 'NFKB1'\n",
       "25. 'NR2F1'\n",
       "26. 'RXRG'\n",
       "27. 'RXRB'\n",
       "28. 'COUP-TFII'\n",
       "29. 'TF65'\n",
       "30. 'NR1H2::RXRA'\n",
       "\n",
       "\n",
       "\n",
       "$`8`\n",
       ":   \n",
       "$`7`\n",
       ":   1. 'ZNF740'\n",
       "2. 'ZNF281'\n",
       "3. 'HOXB5'\n",
       "4. 'ZN281'\n",
       "5. 'ZNF263'\n",
       "6. 'RXRA'\n",
       "7. 'EWSR1-FLI1'\n",
       "8. 'ZBT17'\n",
       "9. 'SP4'\n",
       "10. 'E2F7'\n",
       "11. 'KLF15'\n",
       "12. 'PPARG'\n",
       "13. 'E2F1'\n",
       "14. 'WRNIP1'\n",
       "15. 'ZSC22'\n",
       "16. 'E2F6'\n",
       "17. 'PTF1A'\n",
       "18. 'NKX25'\n",
       "19. 'ETV5'\n",
       "20. 'ZN263'\n",
       "21. 'GTF2I'\n",
       "22. 'ZN467'\n",
       "23. 'MAZ'\n",
       "24. 'VEZF1'\n",
       "25. 'SPIB'\n",
       "26. 'ZN341'\n",
       "27. 'RARA'\n",
       "28. 'PATZ1'\n",
       "29. 'ETS2'\n",
       "30. 'E2F3'\n",
       "31. 'SUH'\n",
       "32. 'TBX3'\n",
       "33. 'INSM1'\n",
       "34. 'NFIC'\n",
       "35. 'MTA3'\n",
       "36. 'KLF12'\n",
       "37. 'ZN257'\n",
       "38. 'LYL1'\n",
       "39. 'ZN816'\n",
       "40. 'TFAP2A'\n",
       "41. 'ZN320'\n",
       "42. 'AP2A'\n",
       "43. 'ZN436'\n",
       "44. 'ZNF467'\n",
       "45. 'SRBP2'\n",
       "46. 'THAP1'\n",
       "47. 'GATA2'\n",
       "48. 'SNAI1'\n",
       "49. 'NR1I3'\n",
       "50. 'ZN563'\n",
       "51. 'SMAD1'\n",
       "52. 'TLX1'\n",
       "\n",
       "\n",
       "\n",
       "$`6`\n",
       ":   1. 'ELF5'\n",
       "2. 'ELF3'\n",
       "3. 'EHF'\n",
       "4. 'GRHL2'\n",
       "5. 'ELF1'\n",
       "6. 'KLF5'\n",
       "7. 'KLF4'\n",
       "8. 'ELF4'\n",
       "9. 'ETV6'\n",
       "10. 'ERG'\n",
       "11. 'SPDEF'\n",
       "12. 'KLF1'\n",
       "13. 'ELF2'\n",
       "14. 'ETV2'\n",
       "15. 'ELK4'\n",
       "16. 'GABPA'\n",
       "17. 'ETV7'\n",
       "18. 'KLF9'\n",
       "19. 'ETS1'\n",
       "20. 'NF-E2'\n",
       "21. 'BACH2'\n",
       "22. 'GRHL1'\n",
       "23. 'SPIC'\n",
       "24. 'ZBTB7A'\n",
       "25. 'PBX3'\n",
       "26. 'JUN'\n",
       "27. 'FOSL1::JUND(var.2)'\n",
       "28. 'ATF4'\n",
       "29. 'TFCP2'\n",
       "30. 'CEBPA'\n",
       "31. 'CEBPB'\n",
       "32. 'IRF8'\n",
       "33. 'CEBPE'\n",
       "34. 'NFIL3'\n",
       "35. 'IRF4'\n",
       "36. 'DDIT3'\n",
       "37. 'CEBP:AP1'\n",
       "\n",
       "\n",
       "\n",
       "$`5`\n",
       ":   1. 'CTCF'\n",
       "2. 'NFYB'\n",
       "3. 'RAD21'\n",
       "4. 'SMC3'\n",
       "\n",
       "\n",
       "\n",
       "$`4`\n",
       ":   1. 'TEAD2'\n",
       "2. 'TEAD4'\n",
       "3. 'TEAD1'\n",
       "4. 'ISL1'\n",
       "5. 'TEAD'\n",
       "6. 'TEAD3'\n",
       "7. 'HOXB3'\n",
       "8. 'ZFHX3'\n",
       "9. 'ESX1'\n",
       "10. 'HOXD3'\n",
       "11. 'PBX2'\n",
       "12. 'HESX1'\n",
       "13. 'GBX1'\n",
       "14. 'HXB7'\n",
       "15. 'HOXB7'\n",
       "16. 'HXA9'\n",
       "\n",
       "\n",
       "\n",
       "$`3`\n",
       ":   1. 'BACH1'\n",
       "2. 'PAX6'\n",
       "3. 'PAX9'\n",
       "4. 'RELA'\n",
       "\n",
       "\n",
       "\n",
       "$`2`\n",
       ":   1. 'TP53'\n",
       "2. 'TP63'\n",
       "3. 'TP73'\n",
       "4. 'BCL11A'\n",
       "5. 'RUNX3'\n",
       "6. 'TFCP2L1'\n",
       "7. 'AHR'\n",
       "8. 'ARNT2'\n",
       "9. 'EGR3'\n",
       "10. 'MAF::NFE2'\n",
       "11. 'DUXA'\n",
       "12. 'TCF7L2'\n",
       "\n",
       "\n",
       "\n",
       "$`1`\n",
       ":   \n",
       "$`0`\n",
       ":   1. 'SPI1'\n",
       "2. 'HOXA4'\n",
       "3. 'PITX1'\n",
       "4. 'PITX2'\n",
       "5. 'PAX5'\n",
       "6. 'POU2F3'\n",
       "7. 'HOXA13'\n",
       "8. 'ONECUT3'\n",
       "9. 'HNF1A'\n",
       "10. 'ZNF384'\n",
       "11. 'RCOR1'\n",
       "12. 'TFAP2D'\n",
       "13. 'NFE2'\n",
       "14. 'FOSB'\n",
       "15. 'BATF::JUN'\n",
       "16. 'SP8'\n",
       "17. 'SMARCC1'\n",
       "18. 'FOS::JUND'\n",
       "19. 'FOSL1::JUN'\n",
       "20. 'FOS::JUN'\n",
       "21. 'FOSL2::JUN'\n",
       "22. 'FOSL1::JUND'\n",
       "23. 'FOSB::JUNB'\n",
       "24. 'FOSL2::JUND'\n",
       "25. 'JUN::JUNB'\n",
       "26. 'FOSL1::JUNB'\n",
       "27. 'FOSL2::JUNB'\n",
       "28. 'FOS::JUNB'\n",
       "29. 'E2F4'\n",
       "30. 'MBD2'\n",
       "31. 'NRF1'\n",
       "32. 'HIC1'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`11`\n",
       "character(0)\n",
       "\n",
       "$`10`\n",
       " [1] \"FOXA2\"     \"DBP\"       \"RFX\"       \"RFX1\"      \"FOXJ2\"     \"RFX7\"     \n",
       " [7] \"SNAI2\"     \"TCF12\"     \"TFE2\"      \"FOXH1\"     \"REST\"      \"ONECUT2\"  \n",
       "[13] \"NHLH1\"     \"BRCA1\"     \"BHA15\"     \"MESP1\"     \"ESRRA\"     \"MYF6\"     \n",
       "[19] \"MSC\"       \"OLIG2\"     \"GLI1\"      \"ZIC1\"      \"ZKSC1\"     \"MYB\"      \n",
       "[25] \"REST-NRSF\" \"MTF1\"      \"ZIC4\"      \"ZSCAN4\"    \"ZIC3\"      \"TBX21\"    \n",
       "[31] \"ESRRB\"     \"NFIA\"      \"CTCFL\"    \n",
       "\n",
       "$`9`\n",
       " [1] \"FOXJ1\"        \"PRDM6\"        \"IRF1\"         \"DMRT1\"        \"ZN770\"       \n",
       " [6] \"IRF9\"         \"FOXJ3\"        \"FOXA1:AR\"     \"STAT1::STAT2\" \"FUBP1\"       \n",
       "[11] \"LEF1\"         \"BPTF\"         \"SMCA5\"        \"ZN121\"        \"IRF3\"        \n",
       "[16] \"IKZF2\"        \"HNF4G\"        \"IRF7\"         \"NFATC2\"       \"HNF4A\"       \n",
       "[21] \"NFATC3\"       \"TCF3\"         \"RBPJ\"         \"NFKB1\"        \"NR2F1\"       \n",
       "[26] \"RXRG\"         \"RXRB\"         \"COUP-TFII\"    \"TF65\"         \"NR1H2::RXRA\" \n",
       "\n",
       "$`8`\n",
       "character(0)\n",
       "\n",
       "$`7`\n",
       " [1] \"ZNF740\"     \"ZNF281\"     \"HOXB5\"      \"ZN281\"      \"ZNF263\"    \n",
       " [6] \"RXRA\"       \"EWSR1-FLI1\" \"ZBT17\"      \"SP4\"        \"E2F7\"      \n",
       "[11] \"KLF15\"      \"PPARG\"      \"E2F1\"       \"WRNIP1\"     \"ZSC22\"     \n",
       "[16] \"E2F6\"       \"PTF1A\"      \"NKX25\"      \"ETV5\"       \"ZN263\"     \n",
       "[21] \"GTF2I\"      \"ZN467\"      \"MAZ\"        \"VEZF1\"      \"SPIB\"      \n",
       "[26] \"ZN341\"      \"RARA\"       \"PATZ1\"      \"ETS2\"       \"E2F3\"      \n",
       "[31] \"SUH\"        \"TBX3\"       \"INSM1\"      \"NFIC\"       \"MTA3\"      \n",
       "[36] \"KLF12\"      \"ZN257\"      \"LYL1\"       \"ZN816\"      \"TFAP2A\"    \n",
       "[41] \"ZN320\"      \"AP2A\"       \"ZN436\"      \"ZNF467\"     \"SRBP2\"     \n",
       "[46] \"THAP1\"      \"GATA2\"      \"SNAI1\"      \"NR1I3\"      \"ZN563\"     \n",
       "[51] \"SMAD1\"      \"TLX1\"      \n",
       "\n",
       "$`6`\n",
       " [1] \"ELF5\"               \"ELF3\"               \"EHF\"               \n",
       " [4] \"GRHL2\"              \"ELF1\"               \"KLF5\"              \n",
       " [7] \"KLF4\"               \"ELF4\"               \"ETV6\"              \n",
       "[10] \"ERG\"                \"SPDEF\"              \"KLF1\"              \n",
       "[13] \"ELF2\"               \"ETV2\"               \"ELK4\"              \n",
       "[16] \"GABPA\"              \"ETV7\"               \"KLF9\"              \n",
       "[19] \"ETS1\"               \"NF-E2\"              \"BACH2\"             \n",
       "[22] \"GRHL1\"              \"SPIC\"               \"ZBTB7A\"            \n",
       "[25] \"PBX3\"               \"JUN\"                \"FOSL1::JUND(var.2)\"\n",
       "[28] \"ATF4\"               \"TFCP2\"              \"CEBPA\"             \n",
       "[31] \"CEBPB\"              \"IRF8\"               \"CEBPE\"             \n",
       "[34] \"NFIL3\"              \"IRF4\"               \"DDIT3\"             \n",
       "[37] \"CEBP:AP1\"          \n",
       "\n",
       "$`5`\n",
       "[1] \"CTCF\"  \"NFYB\"  \"RAD21\" \"SMC3\" \n",
       "\n",
       "$`4`\n",
       " [1] \"TEAD2\" \"TEAD4\" \"TEAD1\" \"ISL1\"  \"TEAD\"  \"TEAD3\" \"HOXB3\" \"ZFHX3\" \"ESX1\" \n",
       "[10] \"HOXD3\" \"PBX2\"  \"HESX1\" \"GBX1\"  \"HXB7\"  \"HOXB7\" \"HXA9\" \n",
       "\n",
       "$`3`\n",
       "[1] \"BACH1\" \"PAX6\"  \"PAX9\"  \"RELA\" \n",
       "\n",
       "$`2`\n",
       " [1] \"TP53\"      \"TP63\"      \"TP73\"      \"BCL11A\"    \"RUNX3\"     \"TFCP2L1\"  \n",
       " [7] \"AHR\"       \"ARNT2\"     \"EGR3\"      \"MAF::NFE2\" \"DUXA\"      \"TCF7L2\"   \n",
       "\n",
       "$`1`\n",
       "character(0)\n",
       "\n",
       "$`0`\n",
       " [1] \"SPI1\"        \"HOXA4\"       \"PITX1\"       \"PITX2\"       \"PAX5\"       \n",
       " [6] \"POU2F3\"      \"HOXA13\"      \"ONECUT3\"     \"HNF1A\"       \"ZNF384\"     \n",
       "[11] \"RCOR1\"       \"TFAP2D\"      \"NFE2\"        \"FOSB\"        \"BATF::JUN\"  \n",
       "[16] \"SP8\"         \"SMARCC1\"     \"FOS::JUND\"   \"FOSL1::JUN\"  \"FOS::JUN\"   \n",
       "[21] \"FOSL2::JUN\"  \"FOSL1::JUND\" \"FOSB::JUNB\"  \"FOSL2::JUND\" \"JUN::JUNB\"  \n",
       "[26] \"FOSL1::JUNB\" \"FOSL2::JUNB\" \"FOS::JUNB\"   \"E2F4\"        \"MBD2\"       \n",
       "[31] \"NRF1\"        \"HIC1\"       \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DefaultAssay(so) <- 'chromvar'\n",
    "TOP5 = list()\n",
    "All = list()\n",
    "for( cl in rev(levels(Idents(so))) ){\n",
    "  message('Testing ', cl,'\\n')\n",
    "  differential.activity = read.csv(file = paste0(outdir, '/clusters/', cl, '_differential.activity_LR.csv'), stringsAsFactors = F)\n",
    "  # Fix homer names\n",
    "  idx = grep('HOMER', differential.activity$X)\n",
    "  differential.activity$TF[ idx ] = gsub('.HOMER', '', differential.activity$X[ idx ])\n",
    "  # Fix hocomoco synonyms\n",
    "  differential.activity$TF2 = differential.activity$TF\n",
    "  idx = differential.activity$TF %in% syn2gen$V1\n",
    "  differential.activity$TF2[ idx ] =  syn2gen$V2[ match(differential.activity$TF[ idx ], syn2gen$V1) ] \n",
    "  differential.activity = differential.activity[ ! duplicated(differential.activity$TF2) , ]\n",
    "    \n",
    "  TOP5[[cl]] = head(differential.activity$X, n = 6) \n",
    "  All[[cl]] = subset(differential.activity, avg_logFC > 1 & p_val_adj < 0.01 )$TF2\n",
    "  if( length(TOP5[[cl]]) > 1 )\n",
    "      for ( tf in TOP5[[cl]] ) {\n",
    "          # Get MotifPlot\n",
    "          p2 = MotifPlot(\n",
    "            object = so,\n",
    "            motifs = gsub('-', '_', tf),\n",
    "            assay = 'peaks'\n",
    "          )\n",
    "          p1 = FeaturePlot(\n",
    "            object = so,\n",
    "            features = tf,\n",
    "            pt.size = 0.1,\n",
    "            cols = c(\"grey\", \"red\"), min.cutoff = 0.5)\n",
    "          P = plot_grid(plotlist = list(p1, p2), ncol = 1)\n",
    "          save_plot(plot = P, filename = paste0(outdir, '/clusters/', cl, '_differential.activity_motifPlot_',tf,'.pdf'), base_width = 4, base_height = 7)\n",
    "       }\n",
    "}\n",
    "\n",
    "All = lapply(All, unique)\n",
    "duplicatedTFs = unlist( All)[ duplicated(unlist( All)) ]\n",
    "All_clusters_specific = lapply(All, setdiff, duplicatedTFs )\n",
    "All_clusters_specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DoHeatmap(so, features = unlist(TOP5), slot = 'data') + scale_fill_gradientn(colors = c(\"grey\", \"red\"))\n",
    "# ggsave(device = 'pdf', file = paste0(outdir, '/general/heatmap_TFmotifs.pdf'), width = 12, height = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing 1\n",
      "\n",
      "\n",
      "Testing 10\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "TF2plot = list('1' = c('E2F4-HUMAN.H11MO.0.A', \"E2F4.MA0470.1\", \"E2F4.HOMER\", 'PGR.HOMER','ESR1.MA0112.3'), \n",
    "             '10' = c('TP73.SwissRegulon', 'FOXJ1.SwissRegulon', \"RFX2-HUMAN.H11MO.0.A\", 'RFX2.MA0600.2'))\n",
    "\n",
    "\n",
    "for( cl in names(TF2plot) ){\n",
    "  message('Testing ', cl,'\\n')\n",
    "  differential.activity = read.csv(file = paste0(outdir, '/clusters/', cl, '_differential.activity_LR.csv'), stringsAsFactors = F)\n",
    "  rownames(differential.activity) = differential.activity$X\n",
    "  for ( tf in TF2plot[[cl]]) {\n",
    "      # Get MotifPlot\n",
    "      p2 = MotifPlot(\n",
    "        object = so,\n",
    "        motifs = gsub('-', '_', tf),\n",
    "        assay = 'peaks'\n",
    "      )\n",
    "      p1 = FeaturePlot(\n",
    "        object = so,\n",
    "        features = tf,\n",
    "        pt.size = 0.2,\n",
    "        cols = c(\"grey\", \"red\"), min.cutoff = 0.5)\n",
    "      P = plot_grid(plotlist = list(p1, p2), ncol = 1)\n",
    "     save_plot(plot = P, filename = paste0(outdir, '/clusters/', cl, '_differential.activity_motifPlot_',tf,'.pdf'), base_width = 4, base_height = 7)\n",
    "   }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding overrepresented TF motifs in DARs\n",
    "\n",
    "Search for DNA motifs that are overrepresented in a set of peaks that are differentially accessible between cell clusters\n",
    "that we previously computed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THERE IS a BUG IN THIS CODE WHEN ADDING MOTIFS APPART from JASPAR\n",
    "# DefaultAssay(so) <- 'peaks'\n",
    "# for( cl in rev(levels(Idents(so))) ){\n",
    "#   cat('Loading ', cl,' DARs\\n')\n",
    "#   da_peaks2 = read.csv(file = paste0(outdir, '/clusters/DARs/', cl, '_LR.csv'), stringsAsFactors = F)\n",
    "#   da_peaks2 = subset(da_peaks2, avg_logFC > 0 & p_val < 0.0001 )\n",
    "#   rownames(da_peaks2) = da_peaks2$X\n",
    "#   enriched.motifs <- FindMotifs(object = so,\n",
    "#     features = rownames(da_peaks2)\n",
    "#   )\n",
    "# #   enriched.motifs = enriched.motifs[ order(enriched.motifs$pvalue, decreasing = F) ,]\n",
    "#   enriched.motifs$cluster = cl\n",
    "# #   # patch \n",
    "# #   enriched.motifs$true_pvalue =   as.numeric(enriched.motifs[ 1 , grep('^pvalue', names(enriched.motifs)) ])\n",
    "# #   enriched.motifs = enriched.motifs[  , - grep('^pvalue', names(enriched.motifs)) ]\n",
    "# #   enriched.motifs = enriched.motifs[order(enriched.motifs$true_pvalue, decreasing = F), ]\n",
    "#   write.csv(enriched.motifs, file = paste0(outdir, '/clusters/DARs/', cl, '_enrichedTFmotifs.csv'))\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
