{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Integration of biopsies and organ donors from 2 studies\n",
    "import anndata, numpy as np, pandas as pd, imp, lpy, scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set some constants\n",
    "sampledir = \"/lustre/scratch117/cellgen/team292/lh20/organoids/\"\n",
    "demul = pd.read_csv(\"demultiplex.csv\",delimiter= '\\t')\n",
    "samplelist = demul[\"sampleID\"]\n",
    "folderlist = [sampledir +x + \"/\" for x in samplelist]\n",
    "soupxfolderlist = [sampledir + x + \"/\" for x in samplelist ] \n",
    "souporcell_genodico = {demul[\"sampleID\"][i] : [demul[\"genotype0\"][i], demul[\"genotype1\"][i], demul[\"genotype2\"][i], demul[\"genotype3\"][i]] for i in range(demul.shape[0])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mCreate AnnData Object\u001b[0m\u001b[34m\n",
      "def createAnnData(folderlist, prefix, souporcell_folderlist = None, souporcell_genodico = None, autoinclude=[\"percent_mito\", \"log2p1_count\", \"n_genes\"], min_cell_per_gene_allowed=3, min_gene_per_cell_allowed=500, sample_obskey = \"sample_names\",doqcplots=False, doinspect=False, mitogeneprefix=\"MT-\", do_log2_normalize=True):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mCreate AnnData Object\\033[0m\\033[34m\"); print(inspect.getsource(createAnnData));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    adatas = []\n",
      "    def pickname(x,y):\n",
      "        if (y == \"singlet\"): return(\"_genotype_\" +str(x))\n",
      "        else: return(\"_doublet\")\n",
      "    def pickname2(x,y,z):\n",
      "        if (y == \"singlet\"): return(z[int(x)])\n",
      "        else: return(\"doublet\")\n",
      "    for i in range(len(folderlist)):\n",
      "        print(\"Processing \" + prefix[i])\n",
      "        adatas.append(sc.read_10x_mtx(folderlist[i]))\n",
      "        if souporcell_folderlist is not None:\n",
      "            try:\n",
      "                res = pd.read_csv(souporcell_folderlist[i] + \"clusters.tsv\",sep='\\t')\n",
      "                if (res.shape[0] != len(adatas[i].obs_names)): raise ValueError( prefix[i] + \" data and soupx barcodes mismatches!\")\n",
      "                adatas[i].obs[\"souporcell\"] = res.status.to_list()\n",
      "                if (souporcell_genodico is None) or (prefix[i] not in souporcell_genodico.keys()):\n",
      "                    adatas[i].obs[\"demultiplexed\"] = [prefix[i] + pickname(x,y)   for x,y in zip( res.assignment.to_list() , res.status.to_list()) ]\n",
      "                else:\n",
      "                    adatas[i].obs[\"demultiplexed\"] = [pickname2(x,y, souporcell_genodico[prefix[i]] )   for x,y in zip( res.assignment.to_list() , res.status.to_list()) ]\n",
      "            except:\n",
      "                print(\"No valid genotyping data for \"+ str(len(adatas[i].obs_names)) + \" cells within \" + prefix[i] + \"! Setting everything to singlets\")\n",
      "                adatas[i].obs[\"souporcell\"] = \"singlet\"\n",
      "                if (souporcell_genodico is None):\n",
      "                    adatas[i].obs[\"demultiplexed\"] = prefix[i] + \"_genotype_0\"\n",
      "                else:\n",
      "                    adatas[i].obs[\"demultiplexed\"] = souporcell_genodico[prefix[i]][0]\n",
      "                    \n",
      "        adatas[i].obs[sample_obskey] = [prefix[i] for hehe in adatas[i].obs_names]\n",
      "        adatas[i].obs_names = [prefix[i] + \"_\" + x for x in adatas[i].obs_names]\n",
      "        if \"n_count\" in autoinclude:\n",
      "            adatas[i].obs['n_count'] = np.sum(adatas[i].X, axis=1).A1\n",
      "        if \"log2p1_count\" in autoinclude:\n",
      "            adatas[i].obs['log2p1_count'] = np.log1p(np.sum(adatas[i].X, axis=1).A1) / math.log(2)\n",
      "        if \"percent_mito\" in autoinclude:\n",
      "            mito_genes = [name for name in adatas[i].var_names if name.startswith(mitogeneprefix)]\n",
      "            adatas[i].obs['percent_mito'] = np.sum(adatas[i][:, mito_genes].X, axis=1).A1 / np.sum(adatas[i].X, axis=1).A1\n",
      "        if \"n_genes\" in autoinclude:\n",
      "            adatas[i].obs['n_genes'] = np.sum(adatas[i].X != 0, axis=1).A1\n",
      "        adatas[i].obs[sample_obskey].value_counts()\n",
      "    adata = mergeAnnData(adatas,index_unique = None)\n",
      "    if min_gene_per_cell_allowed is not None: sc.pp.filter_cells(adata, min_genes=min_gene_per_cell_allowed)\n",
      "    if min_cell_per_gene_allowed is not None: sc.pp.filter_genes(adata, min_cells=min_cell_per_gene_allowed)\n",
      "    \n",
      "    if doqcplots is True:\n",
      "        if \"log2p1_count\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['log2p1_count'], jitter=0.4, groupby=sample_obskey, rotation=90)\n",
      "        if \"n_genes\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['n_genes'], jitter=0.4, groupby=sample_obskey, rotation=90)\n",
      "        if \"percent_mito\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['percent_mito'], jitter=0.4, groupby=sample_obskey, rotation=90)\n",
      "    if do_log2_normalize:\n",
      "        adata.raw = adata.copy()\n",
      "        adata.X = np.log1p(adata.X) / math.log(2)\n",
      "    if \"log2p1_count\" in autoinclude:\n",
      "        adata.obs[\"log2p1_count\"] = pd.to_numeric(adata.obs[\"log2p1_count\"], downcast='float')\n",
      "    if \"percent_mito\" in autoinclude:\n",
      "        adata.obs[\"percent_mito\"] = pd.to_numeric(adata.obs[\"percent_mito\"], downcast='float')\n",
      "    return adata;\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "Processing organoids7090723\n",
      "No valid genotyping data for 4187 cells within organoids7090723! Setting everything to singlets\n",
      "Processing organoids7090724\n",
      "No valid genotyping data for 6220 cells within organoids7090724! Setting everything to singlets\n",
      "Processing organoids7090725\n",
      "No valid genotyping data for 2644 cells within organoids7090725! Setting everything to singlets\n",
      "Processing organoids7090726\n",
      "No valid genotyping data for 2652 cells within organoids7090726! Setting everything to singlets\n",
      "Processing organoids7090727\n",
      "No valid genotyping data for 2804 cells within organoids7090727! Setting everything to singlets\n",
      "Processing organoids7090728\n",
      "No valid genotyping data for 3408 cells within organoids7090728! Setting everything to singlets\n",
      "Processing organoids7090729\n",
      "No valid genotyping data for 3540 cells within organoids7090729! Setting everything to singlets\n",
      "Processing organoids7090730\n",
      "No valid genotyping data for 3657 cells within organoids7090730! Setting everything to singlets\n",
      "Processing organoids7090731\n",
      "No valid genotyping data for 4712 cells within organoids7090731! Setting everything to singlets\n",
      "Processing organoids7090732\n",
      "No valid genotyping data for 4034 cells within organoids7090732! Setting everything to singlets\n",
      "Processing WSSS_END8729879\n",
      "Processing WSSS_END8729880\n",
      "Processing WSSS_END8729881\n",
      "No valid genotyping data for 5984 cells within WSSS_END8729881! Setting everything to singlets\n",
      "Processing WSSS_END8729882\n",
      "Processing WSSS_END8729975\n",
      "Processing WSSS_END8729976\n",
      "Processing WSSS_END8729977\n",
      "Processing WSSS_END8729978\n",
      "Processing WSSS_END8729979\n",
      "Processing WSSS_END8729980\n",
      "Processing WSSS_END8729981\n",
      "Processing WSSS_END8729982\n",
      "Processing WSSS_END8729983\n",
      "Processing WSSS_END8729984\n",
      "Processing WSSS_END8729985\n",
      "Processing WSSS_END8730071\n",
      "Processing WSSS_END8730167\n",
      "Processing WSSS_END8730168\n",
      "Processing WSSS_END8730169\n",
      "Processing WSSS_END8730170\n",
      "Processing WSSS_END8730171\n",
      "Processing WSSS_END8730172\n",
      "Processing WSSS_END8730173\n",
      "Processing WSSS_END8730174\n",
      "Processing WSSS_END8730263\n",
      "Processing WSSS_END8730264\n",
      "Processing WSSS_END8730359\n",
      "Processing WSSS_END8730455\n",
      "Processing WSSS_END8730456\n",
      "Processing WSSS_END8730457\n",
      "Processing WSSS_END8730458\n",
      "Processing WSSS_END8730459\n",
      "Processing WSSS_END8730460\n",
      "Processing WSSS_END8730461\n",
      "Processing WSSS_END8730462\n",
      "Processing WSSS_END8730551\n",
      "Processing WSSS_END8730552\n",
      "Processing WSSS_END9263313\n",
      "Processing WSSS_END9263314\n",
      "Processing WSSS_END9263315\n",
      "Processing WSSS_END9263316\n",
      "Processing WSSS_END9263317\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'souporcell' as categorical\n",
      "... storing 'demultiplexed' as categorical\n",
      "... storing 'sample_names' as categorical\n"
     ]
    }
   ],
   "source": [
    "#Load data\n",
    "adata=lpy.createAnnData( folderlist, samplelist, souporcell_folderlist = soupxfolderlist, souporcell_genodico=souporcell_genodico, doqcplots=True, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mAdd metadata from sample table to anndata\u001b[0m\u001b[34m\n",
      "def addMetadata(adata, metadata, obs_key, meta_key, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mAdd metadata from sample table to anndata\\033[0m\\033[34m\"); print(inspect.getsource(addMetadata));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    aslist = metadata[meta_key].tolist()\n",
      "    rowmap = {i : aslist.index(i)  for i in aslist}\n",
      "    for val in metadata.columns:\n",
      "        if val != meta_key:\n",
      "            aslist = metadata[val].tolist()\n",
      "            adata.obs[val] = [aslist[rowmap[i]] for i in adata.obs[obs_key] ]\n",
      "    return adata;\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "meta = pd.read_csv(\"samplemeta.csv\",delimiter= '\\t')\n",
    "adata = lpy.addMetadata(adata,meta, \"sample_names\",\"sample_names\",doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "WSSS_END9263316     10269\n",
       "WSSS_END9263314      7581\n",
       "WSSS_END9263315      6892\n",
       "organoids7090724     6220\n",
       "WSSS_END8730071      6132\n",
       "WSSS_END8729881      5981\n",
       "WSSS_END9263313      5889\n",
       "WSSS_END8730359      4946\n",
       "organoids7090731     4711\n",
       "WSSS_END8730461      4411\n",
       "WSSS_END8730456      4381\n",
       "organoids7090723     4187\n",
       "organoids7090732     4032\n",
       "WSSS_END8729984      3836\n",
       "WSSS_END8729882      3758\n",
       "WSSS_END8729880      3700\n",
       "WSSS_END8730551      3685\n",
       "organoids7090730     3657\n",
       "organoids7090729     3540\n",
       "organoids7090728     3408\n",
       "WSSS_END8729985      3370\n",
       "WSSS_END8730458      3341\n",
       "WSSS_END8729979      3332\n",
       "WSSS_END8730457      3313\n",
       "WSSS_END8730171      3224\n",
       "WSSS_END8730552      3081\n",
       "WSSS_END8730460      3059\n",
       "WSSS_END8730173      2949\n",
       "WSSS_END8730174      2933\n",
       "WSSS_END8729982      2929\n",
       "WSSS_END8730167      2826\n",
       "organoids7090727     2803\n",
       "WSSS_END8730170      2787\n",
       "WSSS_END8729981      2772\n",
       "WSSS_END8729976      2721\n",
       "organoids7090726     2652\n",
       "organoids7090725     2644\n",
       "WSSS_END8729980      2632\n",
       "WSSS_END8730172      2584\n",
       "WSSS_END8730459      2528\n",
       "WSSS_END8729975      2528\n",
       "WSSS_END8730455      2522\n",
       "WSSS_END8729977      2357\n",
       "WSSS_END8729983      2225\n",
       "WSSS_END8730169      2024\n",
       "WSSS_END8729978      2003\n",
       "WSSS_END8729879      1956\n",
       "WSSS_END8730263      1938\n",
       "WSSS_END8730168      1932\n",
       "WSSS_END8730264      1914\n",
       "WSSS_END8730462      1735\n",
       "WSSS_END9263317       699\n",
       "Name: sample_names, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"sample_names\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mDetect Doublets and defining cells to filter\u001b[0m\u001b[34m\n",
      "def scrub(adata, batch_obsattrib, bonf_threshold = 0.01, add_qc_metrics=False,mito_prefix= \"MT-\", obskey_cellfilter = \"filtered_cells\", add_cell_filter={\"max_percent_mito\": 0.15, \"scrublet_local_pred\": False}, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mDetect Doublets and defining cells to filter\\033[0m\\033[34m\"); print(inspect.getsource(scrub));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    \n",
      "    import scrublet as scr\n",
      "    import scanpy as sc\n",
      "    print(\"spliting data using attribute \" + batch_obsattrib)\n",
      "    adatas = splitAnnData(adata, batch_obsattrib)\n",
      "\n",
      "    if (add_qc_metrics):\n",
      "        mito_genes = [name for name in adata.var_names if name.startswith(mito_prefix)]\n",
      "        adata.obs['log2p1_RNA_count'] = np.log1p(adata.X.sum(axis=1).A1) / math.log(2)\n",
      "        adata.obs['percent_mito'] = np.sum(adata[:, mito_genes].X, axis=1).A1 / np.sum(adata.X, axis=1).A1\n",
      "\n",
      "    dalist = list(adata.obs_names)\n",
      "    tmap = {}\n",
      "    for i in range(len(adata.obs_names)):\n",
      "        tmap.update( {adata.obs_names[i] : i})\n",
      "    adata.obs[\"scrublet_pred\"] = False\n",
      "    adata.obs[\"scrublet_local_pred\"] = False\n",
      "    adata.obs[\"scrublet_score\"] = 0.0\n",
      "    adata.obs[\"scrublet_cluster_score\"] = 0.0\n",
      "\n",
      "    # Luz's double clustering approach\n",
      "    for i in range(len(adatas)):\n",
      "\n",
      "        print(\"processing \" + str(i) + \"/\" + str(len(adatas)))\n",
      "        curscr = scr.Scrublet(adatas[i].X)\n",
      "        doublet_scores, predicted_doublets = curscr.scrub_doublets()\n",
      "        adatas[i].obs['scrublet_score'] = doublet_scores\n",
      "        adatas[i].obs['scrublet_pred'] = predicted_doublets\n",
      "        #overcluster prep. run turbo basic scanpy pipeline\n",
      "        sc.pp.normalize_per_cell(adatas[i], counts_per_cell_after=1e4)\n",
      "        sc.pp.log1p(adatas[i])\n",
      "        sc.pp.highly_variable_genes(adatas[i], min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
      "        adatas[i] = adatas[i][:, adatas[i].var['highly_variable']]\n",
      "        sc.pp.scale(adatas[i], max_value=10)\n",
      "        sc.tl.pca(adatas[i], svd_solver='arpack')\n",
      "        sc.pp.neighbors(adatas[i])\n",
      "        #overclustering proper - do basic clustering first, then cluster each cluster\n",
      "        sc.tl.leiden(adatas[i])\n",
      "        adatas[i].obs['leiden'] = [str(i) for i in adatas[i].obs['leiden']]\n",
      "        for clus in np.unique(adatas[i].obs['leiden']):\n",
      "            adata_sub = adatas[i][adatas[i].obs['leiden']==clus].copy()\n",
      "            sc.tl.leiden(adata_sub)\n",
      "            adata_sub.obs['leiden'] = [clus+','+i for i in adata_sub.obs['leiden']]\n",
      "            adatas[i].obs.loc[adata_sub.obs_names,'leiden'] = adata_sub.obs['leiden']\n",
      "\n",
      "        #compute the cluster scores - the median of Scrublet scores per overclustered cluster\n",
      "        for clus in np.unique(adatas[i].obs['leiden']):\n",
      "            adatas[i].obs.loc[adatas[i].obs['leiden']==clus, 'scrublet_cluster_score'] = \\\n",
      "                np.median(adatas[i].obs.loc[adatas[i].obs['leiden']==clus, 'scrublet_score'])\n",
      "        #now compute doublet p-values. figure out the median and mad (from above-median values) for the distribution\n",
      "        med = np.median(adatas[i].obs['scrublet_cluster_score'])\n",
      "        mask = adatas[i].obs['scrublet_cluster_score']>med\n",
      "        mad = np.median(adatas[i].obs['scrublet_cluster_score'][mask]-med)\n",
      "        #let's do a one-sided test. the Bertie write-up does not address this but it makes sense\n",
      "        zscores = (adatas[i].obs['scrublet_cluster_score'].values - med) / (1.4826 * mad)\n",
      "        #adatas[i].obs['zscore'] = zscores\n",
      "        pvals = 1-scipy.stats.norm.cdf(zscores)\n",
      "        #adatas[i].obs['bh_pval'] = bh(pvals)\n",
      "        pvals = bonf(pvals)\n",
      "        adatas[i].obs['scrublet_local_pred'] = pvals < bonf_threshold\n",
      "        map = [tmap[s] for s in adatas[i].obs_names]\n",
      "        print(\"annoying values\")\n",
      "        for j in range(len(adatas[i].obs_names)):\n",
      "            adata.obs[\"scrublet_pred\" ][map[j]] = adatas[i].obs[\"scrublet_pred\"][j]\n",
      "            adata.obs[\"scrublet_local_pred\"][map[j]] = adatas[i].obs[\"scrublet_local_pred\"][j]\n",
      "            adata.obs[\"scrublet_cluster_score\"][map[j]] =  adatas[i].obs[\"scrublet_cluster_score\"][j]\n",
      "            adata.obs[\"scrublet_score\"][map[j]] =  pvals[j]\n",
      "\n",
      "    adata.obs[\"scrublet_cluster_score\"] = pd.to_numeric(adata.obs[\"scrublet_cluster_score\"], downcast='float')\n",
      "    adata.obs[\"scrublet_score\"] = pd.to_numeric(adata.obs[\"scrublet_score\"], downcast='float')\n",
      "\n",
      "    if add_cell_filter is not None:\n",
      "        curflt = np.zeros(adata.obs.shape[0], dtype=bool) \n",
      "        for k,v in add_cell_filter.items():\n",
      "            if k == \"max_percent_mito\":\n",
      "                print(str(sum(adata.obs['percent_mito'] > v)) + \" cells filtered by mito threshold\")\n",
      "                curflt |= adata.obs['percent_mito'] > v\n",
      "            else:\n",
      "                print(str(sum(adata.obs[k] != v)) + \" cells filtered by \" + k + \" == \" + str(v) + \" criterion\")\n",
      "                curflt |= (adata.obs[k] != v)\n",
      "        adata.obs[obskey_cellfilter] = curflt\n",
      "    return adata\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "spliting data using attribute sample_names\n",
      "processing 0/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 4.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.9%\n",
      "Elapsed time: 13.0 seconds\n",
      "annoying values\n",
      "processing 1/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 10.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.4%\n",
      "Elapsed time: 2.2 seconds\n",
      "annoying values\n",
      "processing 2/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.78\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 9.9 seconds\n",
      "annoying values\n",
      "processing 3/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 16.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.3%\n",
      "Elapsed time: 4.3 seconds\n",
      "annoying values\n",
      "processing 4/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 10.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.4%\n",
      "Elapsed time: 1.6 seconds\n",
      "annoying values\n",
      "processing 5/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 14.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.4%\n",
      "Elapsed time: 1.7 seconds\n",
      "annoying values\n",
      "processing 6/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.73\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 7.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.5%\n",
      "Elapsed time: 5.8 seconds\n",
      "annoying values\n",
      "processing 7/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 7.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.1%\n",
      "Elapsed time: 2.7 seconds\n",
      "annoying values\n",
      "processing 8/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 7.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.9%\n",
      "Elapsed time: 1.5 seconds\n",
      "annoying values\n",
      "processing 9/52\n",
      "Preprocessing...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 7.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.9%\n",
      "Elapsed time: 2.1 seconds\n",
      "annoying values\n",
      "processing 10/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 7.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.9%\n",
      "Elapsed time: 2.2 seconds\n",
      "annoying values\n",
      "processing 11/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 8.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.2%\n",
      "Elapsed time: 2.3 seconds\n",
      "annoying values\n",
      "processing 12/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.69\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 3.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.2%\n",
      "Elapsed time: 2.7 seconds\n",
      "annoying values\n",
      "processing 13/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 17.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.7%\n",
      "Elapsed time: 3.3 seconds\n",
      "annoying values\n",
      "processing 14/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.71\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 6.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 4.1 seconds\n",
      "annoying values\n",
      "processing 15/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.51\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.2%\n",
      "Elapsed time: 2.9 seconds\n",
      "annoying values\n",
      "processing 16/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 10.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.9%\n",
      "Elapsed time: 2.0 seconds\n",
      "annoying values\n",
      "processing 17/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 6.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.6%\n",
      "Elapsed time: 2.4 seconds\n",
      "annoying values\n",
      "processing 18/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 8.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.7%\n",
      "Elapsed time: 2.5 seconds\n",
      "annoying values\n",
      "processing 19/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 10.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.6%\n",
      "Elapsed time: 2.2 seconds\n",
      "annoying values\n",
      "processing 20/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 1.4%\n",
      "Estimated detectable doublet fraction = 35.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.0%\n",
      "Elapsed time: 3.0 seconds\n",
      "annoying values\n",
      "processing 21/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 12.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.2%\n",
      "Elapsed time: 1.5 seconds\n",
      "annoying values\n",
      "processing 22/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 7.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.0%\n",
      "Elapsed time: 1.6 seconds\n",
      "annoying values\n",
      "processing 23/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 34.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.3%\n",
      "Elapsed time: 3.4 seconds\n",
      "annoying values\n",
      "processing 24/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 7.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.1%\n",
      "Elapsed time: 1.7 seconds\n",
      "annoying values\n",
      "processing 25/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 12.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.5%\n",
      "Elapsed time: 2.8 seconds\n",
      "annoying values\n",
      "processing 26/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 2.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.1%\n",
      "Elapsed time: 6.8 seconds\n",
      "annoying values\n",
      "processing 27/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 2.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 7.0 seconds\n",
      "annoying values\n",
      "processing 28/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 16.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.7%\n",
      "Elapsed time: 2.1 seconds\n",
      "annoying values\n",
      "processing 29/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 39.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.4%\n",
      "Elapsed time: 2.8 seconds\n",
      "annoying values\n",
      "processing 30/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.66\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 9.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 2.5 seconds\n",
      "annoying values\n",
      "processing 31/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.72\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 5.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.6%\n",
      "Elapsed time: 5.7 seconds\n",
      "annoying values\n",
      "processing 32/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 10.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.2%\n",
      "Elapsed time: 3.5 seconds\n",
      "annoying values\n",
      "processing 33/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 12.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 2.4 seconds\n",
      "annoying values\n",
      "processing 34/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.70\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 11.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 3.6 seconds\n",
      "annoying values\n",
      "processing 35/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 16.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.2%\n",
      "Elapsed time: 2.4 seconds\n",
      "annoying values\n",
      "processing 36/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.69\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 11.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.7%\n",
      "Elapsed time: 3.9 seconds\n",
      "annoying values\n",
      "processing 37/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 13.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 1.5 seconds\n",
      "annoying values\n",
      "processing 38/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.70\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 6.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.0%\n",
      "Elapsed time: 3.5 seconds\n",
      "annoying values\n",
      "processing 39/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 5.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.8%\n",
      "Elapsed time: 1.9 seconds\n",
      "annoying values\n",
      "processing 40/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 41.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.1%\n",
      "Elapsed time: 2.1 seconds\n",
      "annoying values\n",
      "processing 41/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.73\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 2.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.2%\n",
      "Elapsed time: 5.1 seconds\n",
      "annoying values\n",
      "processing 42/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 9.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.3%\n",
      "Elapsed time: 2.8 seconds\n",
      "annoying values\n",
      "processing 43/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.68\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 5.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.5%\n",
      "Elapsed time: 3.2 seconds\n",
      "annoying values\n",
      "processing 44/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 13.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.6%\n",
      "Elapsed time: 2.0 seconds\n",
      "annoying values\n",
      "processing 45/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 14.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.9%\n",
      "Elapsed time: 0.5 seconds\n",
      "annoying values\n",
      "processing 46/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 8.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.9%\n",
      "Elapsed time: 2.0 seconds\n",
      "annoying values\n",
      "processing 47/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.73\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 4.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.6%\n",
      "Elapsed time: 5.6 seconds\n",
      "annoying values\n",
      "processing 48/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 6.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.0%\n",
      "Elapsed time: 2.2 seconds\n",
      "annoying values\n",
      "processing 49/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 8.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.5%\n",
      "Elapsed time: 2.0 seconds\n",
      "annoying values\n",
      "processing 50/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 8.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.2%\n",
      "Elapsed time: 2.5 seconds\n",
      "annoying values\n",
      "processing 51/52\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 9.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.6%\n",
      "Elapsed time: 1.5 seconds\n",
      "annoying values\n",
      "5413 cells filtered by souporcell == singlet criterion\n",
      "11987 cells filtered by mito threshold\n",
      "865 cells filtered by scrublet_local_pred == False criterion\n"
     ]
    }
   ],
   "source": [
    "obskey_filteredcells = \"filtered_cells\"\n",
    "adata=lpy.scrub(adata, \"sample_names\",obskey_cellfilter= obskey_filteredcells,add_cell_filter={\"souporcell\": \"singlet\" ,\"max_percent_mito\":0.3, \"scrublet_local_pred\": False }, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mPrepare Data for Scvi/TotalVi\u001b[0m\u001b[34m\n",
      "def scvi_prepare(anndatapath, field, cellfilter = None, nbgenes = 5000, genes_to_filter= None, use_ccfilter_prefix=None, citeseqkey = \"protein_expression\", use_raw_slot_instead =None, min_cell_threshold= 0, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mPrepare Data for Scvi/TotalVi\\033[0m\\033[34m\"); print(inspect.getsource(scvi_prepare));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    if use_ccfilter_prefix is not None :\n",
      "        genes_to_filter = [use_ccfilter_prefix + x for x in [\"HMGB2\",\"CDK1\",\"NUSAP1\",\"UBE2C\",\"BIRC5\",\"TPX2\",\"TOP2A\",\"NDC80\",\"CKS2\",\"NUF2\",\"CKS1B\",\"MKI67\",\"TMPO\",\"CENPF\",\"TACC3\",\"FAM64A\",\"SMC4\",\"CCNB2\",\"CKAP2L\",\"CKAP2\",\"AURKB\",\"BUB1\",\"KIF11\",\"ANP32E\",\"TUBB4B\",\"GTSE1\",\"KIF20B\",\"HJURP\",\"CDCA3\",\"HN1\",\"CDC20\",\"TTK\",\"CDC25C\",\"KIF2C\",\"RANGAP1\",\"NCAPD2\",\"DLGAP5\",\"CDCA2\",\"CDCA8\",\"ECT2\",\"KIF23\",\"HMMR\",\"AURKA\",\"PSRC1\",\"ANLN\",\"LBR\",\"CKAP5\",\"CENPE\",\"CTCF\",\"NEK2\",\"G2E3\",\"GAS2L3\",\"CBX5\",\"CENPA\",\"MCM5\",\"PCNA\",\"TYMS\",\"FEN1\",\"MCM2\",\"MCM4\",\"RRM1\",\"UNG\",\"GINS2\",\"MCM6\",\"CDCA7\",\"DTL\",\"PRIM1\",\"UHRF1\",\"MLF1IP\",\"HELLS\",\"RFC2\",\"RPA2\",\"NASP\",\"RAD51AP1\",\"GMNN\",\"WDR76\",\"SLBP\",\"CCNE2\",\"UBR7\",\"POLD3\",\"MSH2\",\"ATAD2\",\"RAD51\",\"RRM2\",\"CDC45\",\"CDC6\",\"EXO1\",\"TIPIN\",\"DSCC1\",\"BLM\",\"CASP8AP2\",\"USP1\",\"CLSPN\",\"POLA1\",\"CHAF1B\",\"BRIP1\",\"E2F8\"]]\n",
      "    if (isinstance(anndatapath, str)): cite = anndata.read_h5ad(anndatapath)\n",
      "    else: cite = anndatapath\n",
      "    if use_raw_slot_instead is None: # default behavior, check for the existance of the raw slot\n",
      "        use_raw_slot_instead = cite.raw is not None\n",
      "    \n",
      "    split = splitAnnData(cite, field, entryfilter= cellfilter, getnames = True, use_raw_slot_instead= use_raw_slot_instead,min_cell_threshold=min_cell_threshold)\n",
      "    if genes_to_filter is not None:\n",
      "        for i in range(len(split[\"datalist\"])):\n",
      "            split[\"datalist\"][i] = split[\"datalist\"][i][:, [g not in genes_to_filter for g in split[\"datalist\"][i].var_names]]\n",
      "    \n",
      "    \n",
      "\n",
      "    dataset = GeneExpressionDataset()\n",
      "    if citeseqkey in cite.obsm.keys():\n",
      "        if citeseqkey is not \"protein_expression\":\n",
      "            for i in range(len(split[\"datalist\"])):\n",
      "                split[\"datalist\"][i].obsm[\"protein_expression\"] = split[\"datalist\"][i].obsm[citeseqkey]\n",
      "        dataset.populate_from_datasets([AnnDatasetFromAnnData(ad=s,cell_measurements_col_mappings={\"protein_expression\":\"protein_names\"})   for s in split[\"datalist\"] ])\n",
      "    else:\n",
      "        dataset.populate_from_datasets([AnnDatasetFromAnnData(ad=s)   for s in split[\"datalist\"] ])\n",
      "\n",
      "    if nbgenes != 0:\n",
      "        dataset.subsample_genes(nbgenes)\n",
      "    return{\"dataset\":dataset , \"names\" : split[\"ordered\"]}\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "[2020-11-02 12:22:58,608] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:22:58,854] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:22:58,856] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:22:59,441] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:22:59,525] INFO - scvi.dataset.dataset | Downsampled from 1860 to 1860 cells\n",
      "[2020-11-02 12:22:59,534] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:22:59,819] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:22:59,821] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:00,515] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:00,613] INFO - scvi.dataset.dataset | Downsampled from 2185 to 2185 cells\n",
      "[2020-11-02 12:23:00,806] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:00,808] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:01,266] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:01,342] INFO - scvi.dataset.dataset | Downsampled from 9210 to 9210 cells\n",
      "[2020-11-02 12:23:01,347] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:01,842] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:01,844] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:03,039] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:03,216] INFO - scvi.dataset.dataset | Downsampled from 3941 to 3941 cells\n",
      "[2020-11-02 12:23:03,231] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:03,475] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:03,477] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:04,078] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:04,160] INFO - scvi.dataset.dataset | Downsampled from 1836 to 1836 cells\n",
      "[2020-11-02 12:23:04,169] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:04,409] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:04,411] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:05,000] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:05,080] INFO - scvi.dataset.dataset | Downsampled from 1788 to 1788 cells\n",
      "[2020-11-02 12:23:05,088] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:05,834] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:05,836] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:07,544] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:07,811] INFO - scvi.dataset.dataset | Downsampled from 5944 to 5944 cells\n",
      "[2020-11-02 12:23:07,831] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:08,230] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:08,232] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:09,195] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:09,337] INFO - scvi.dataset.dataset | Downsampled from 3146 to 3146 cells\n",
      "[2020-11-02 12:23:09,349] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:09,582] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:09,584] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:10,119] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:10,197] INFO - scvi.dataset.dataset | Downsampled from 1745 to 1745 cells\n",
      "[2020-11-02 12:23:10,205] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:10,564] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:10,566] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:11,389] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:11,519] INFO - scvi.dataset.dataset | Downsampled from 2773 to 2773 cells\n",
      "[2020-11-02 12:23:11,530] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:11,849] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:11,851] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:12,538] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:12,641] INFO - scvi.dataset.dataset | Downsampled from 2303 to 2303 cells\n",
      "[2020-11-02 12:23:12,651] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:12,971] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:12,973] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:13,737] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 12:23:13,849] INFO - scvi.dataset.dataset | Downsampled from 2486 to 2486 cells\n",
      "[2020-11-02 12:23:13,859] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:14,239] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:14,241] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:15,130] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:15,264] INFO - scvi.dataset.dataset | Downsampled from 2988 to 2988 cells\n",
      "[2020-11-02 12:23:15,277] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:15,728] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:15,730] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:16,836] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:16,996] INFO - scvi.dataset.dataset | Downsampled from 3563 to 3563 cells\n",
      "[2020-11-02 12:23:17,010] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:17,616] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:17,618] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:18,987] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:19,203] INFO - scvi.dataset.dataset | Downsampled from 4815 to 4815 cells\n",
      "[2020-11-02 12:23:19,221] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:19,669] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:19,671] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:20,680] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:20,839] INFO - scvi.dataset.dataset | Downsampled from 3547 to 3547 cells\n",
      "[2020-11-02 12:23:20,853] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:21,141] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:21,143] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:21,811] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:21,909] INFO - scvi.dataset.dataset | Downsampled from 2181 to 2181 cells\n",
      "[2020-11-02 12:23:21,919] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:22,210] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:22,212] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:22,905] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:23,005] INFO - scvi.dataset.dataset | Downsampled from 2223 to 2223 cells\n",
      "[2020-11-02 12:23:23,015] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:23,371] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:23,373] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:24,196] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:24,320] INFO - scvi.dataset.dataset | Downsampled from 2761 to 2761 cells\n",
      "[2020-11-02 12:23:24,331] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:24,593] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:24,595] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:25,209] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:25,297] INFO - scvi.dataset.dataset | Downsampled from 1968 to 1968 cells\n",
      "[2020-11-02 12:23:25,306] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:25,650] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:25,652] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:26,466] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:26,585] INFO - scvi.dataset.dataset | Downsampled from 2652 to 2652 cells\n",
      "[2020-11-02 12:23:26,597] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:26,790] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:26,792] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:27,234] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:27,296] INFO - scvi.dataset.dataset | Downsampled from 1399 to 1399 cells\n",
      "[2020-11-02 12:23:27,304] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:27,533] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:27,535] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:28,070] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:28,145] INFO - scvi.dataset.dataset | Downsampled from 1676 to 1676 cells\n",
      "[2020-11-02 12:23:28,153] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:28,646] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:28,648] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:29,737] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:29,912] INFO - scvi.dataset.dataset | Downsampled from 3899 to 3899 cells\n",
      "[2020-11-02 12:23:29,927] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:30,168] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:30,169] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:30,735] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:30,815] INFO - scvi.dataset.dataset | Downsampled from 1790 to 1790 cells\n",
      "[2020-11-02 12:23:30,824] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:31,240] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:31,242] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:32,233] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:32,377] INFO - scvi.dataset.dataset | Downsampled from 3211 to 3211 cells\n",
      "[2020-11-02 12:23:32,390] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:33,161] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:33,163] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:34,953] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:35,228] INFO - scvi.dataset.dataset | Downsampled from 6137 to 6137 cells\n",
      "[2020-11-02 12:23:35,250] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:36,106] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:36,108] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:38,041] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:38,348] INFO - scvi.dataset.dataset | Downsampled from 6849 to 6849 cells\n",
      "[2020-11-02 12:23:38,372] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:38,715] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:38,717] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:39,521] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 12:23:39,638] INFO - scvi.dataset.dataset | Downsampled from 2632 to 2632 cells\n",
      "[2020-11-02 12:23:39,649] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:40,074] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:40,076] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:41,050] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:41,198] INFO - scvi.dataset.dataset | Downsampled from 3292 to 3292 cells\n",
      "[2020-11-02 12:23:41,211] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:41,605] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:41,607] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:42,521] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:42,657] INFO - scvi.dataset.dataset | Downsampled from 2983 to 2983 cells\n",
      "[2020-11-02 12:23:42,670] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:43,453] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:43,455] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:45,262] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:45,506] INFO - scvi.dataset.dataset | Downsampled from 5415 to 5415 cells\n",
      "[2020-11-02 12:23:45,525] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:45,983] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:45,985] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:47,101] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:47,262] INFO - scvi.dataset.dataset | Downsampled from 3599 to 3599 cells\n",
      "[2020-11-02 12:23:47,276] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:47,619] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:47,621] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:48,418] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:48,534] INFO - scvi.dataset.dataset | Downsampled from 2600 to 2600 cells\n",
      "[2020-11-02 12:23:48,545] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:49,074] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:49,076] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:50,347] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:50,535] INFO - scvi.dataset.dataset | Downsampled from 4183 to 4183 cells\n",
      "[2020-11-02 12:23:50,551] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:50,937] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:50,939] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:51,822] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:51,957] INFO - scvi.dataset.dataset | Downsampled from 3004 to 3004 cells\n",
      "[2020-11-02 12:23:51,969] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:52,484] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:52,486] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:53,693] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:53,876] INFO - scvi.dataset.dataset | Downsampled from 4086 to 4086 cells\n",
      "[2020-11-02 12:23:53,892] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:54,132] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:54,134] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:54,693] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:54,773] INFO - scvi.dataset.dataset | Downsampled from 1785 to 1785 cells\n",
      "[2020-11-02 12:23:54,782] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:55,373] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:55,375] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:56,669] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:56,883] INFO - scvi.dataset.dataset | Downsampled from 4675 to 4675 cells\n",
      "[2020-11-02 12:23:56,900] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:57,143] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:57,145] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:57,717] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:57,800] INFO - scvi.dataset.dataset | Downsampled from 1801 to 1801 cells\n",
      "[2020-11-02 12:23:57,808] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:58,147] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:58,149] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:23:58,935] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:23:59,053] INFO - scvi.dataset.dataset | Downsampled from 2586 to 2586 cells\n",
      "[2020-11-02 12:23:59,065] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:23:59,789] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:23:59,791] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:01,465] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:01,727] INFO - scvi.dataset.dataset | Downsampled from 5828 to 5828 cells\n",
      "[2020-11-02 12:24:01,748] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:02,096] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:02,098] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:02,921] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:03,040] INFO - scvi.dataset.dataset | Downsampled from 2624 to 2624 cells\n",
      "[2020-11-02 12:24:03,053] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:03,503] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:03,505] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:04,552] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:04,712] INFO - scvi.dataset.dataset | Downsampled from 3496 to 3496 cells\n",
      "[2020-11-02 12:24:04,726] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:05,025] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:05,027] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:05,721] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:05,819] INFO - scvi.dataset.dataset | Downsampled from 2211 to 2211 cells\n",
      "[2020-11-02 12:24:05,829] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:05,934] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:05,936] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:06,136] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 12:24:06,165] INFO - scvi.dataset.dataset | Downsampled from 629 to 629 cells\n",
      "[2020-11-02 12:24:06,170] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:06,451] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:06,453] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:07,084] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:07,181] INFO - scvi.dataset.dataset | Downsampled from 2122 to 2122 cells\n",
      "[2020-11-02 12:24:07,190] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:07,960] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:07,962] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:09,781] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:10,061] INFO - scvi.dataset.dataset | Downsampled from 6208 to 6208 cells\n",
      "[2020-11-02 12:24:10,083] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:10,419] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:10,421] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:11,210] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:11,326] INFO - scvi.dataset.dataset | Downsampled from 2590 to 2590 cells\n",
      "[2020-11-02 12:24:11,337] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:11,558] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:11,560] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:12,065] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:12,139] INFO - scvi.dataset.dataset | Downsampled from 1629 to 1629 cells\n",
      "[2020-11-02 12:24:12,147] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:12,595] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:12,597] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:13,599] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:13,758] INFO - scvi.dataset.dataset | Downsampled from 3526 to 3526 cells\n",
      "[2020-11-02 12:24:13,771] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 12:24:13,971] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:13,973] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:14,460] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:14,525] INFO - scvi.dataset.dataset | Downsampled from 1447 to 1447 cells\n",
      "[2020-11-02 12:24:14,532] INFO - scvi.dataset.dataset | Merging datasets. Input objects are modified in place.\n",
      "[2020-11-02 12:24:14,534] INFO - scvi.dataset.dataset | Gene names and cell measurement names are assumed to have a non-null intersection between datasets.\n",
      "[2020-11-02 12:24:15,600] INFO - scvi.dataset.dataset | Keeping 27569 genes\n",
      "[2020-11-02 12:24:16,232] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:16,460] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:16,462] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:17,188] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:17,449] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:17,451] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:17,989] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:18,071] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:18,073] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:19,333] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:19,848] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:19,850] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:20,481] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:20,706] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:20,708] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:21,323] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:21,544] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:21,545] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:23,361] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:24,073] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:24,075] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:25,084] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:25,470] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:25,472] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:26,043] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:26,255] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:26,257] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:27,127] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:27,451] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:27,453] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:28,190] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:28,470] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:28,472] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:29,282] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:29,588] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:29,590] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:30,516] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:30,881] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:30,883] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:32,036] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:32,450] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:32,452] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:33,915] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:34,489] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:34,491] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:35,579] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:36,003] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:36,005] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:36,718] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:36,982] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:36,984] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:37,718] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:37,989] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:37,991] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 12:24:38,867] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:39,195] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:39,197] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:39,847] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:40,089] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:40,091] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:40,948] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:41,271] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:41,273] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:41,746] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:41,912] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:41,914] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:42,480] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:42,687] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:42,689] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:43,866] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:44,321] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:44,323] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:44,921] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:45,141] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:45,143] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:46,182] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:46,564] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:46,566] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:48,466] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:49,199] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:49,201] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:51,269] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:52,092] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:52,094] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:52,939] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:53,260] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:53,262] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:54,295] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:54,697] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:54,699] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:55,659] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:56,015] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:56,017] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:57,759] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:24:58,400] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:24:58,402] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:24:59,608] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:00,040] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:00,042] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:00,890] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:01,209] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:01,211] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:02,542] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:03,042] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:03,044] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:03,993] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:04,365] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:04,367] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:05,646] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:06,145] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:06,147] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:06,742] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:06,955] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:06,957] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:08,348] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:08,904] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:08,906] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:09,511] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:09,725] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:09,727] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:10,554] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:10,868] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:10,870] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:12,676] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:13,378] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:13,380] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:14,236] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:14,609] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:14,611] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:15,716] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:16,137] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:16,139] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:16,874] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:17,139] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:17,141] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:17,372] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:17,456] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:17,458] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:18,139] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:18,401] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:18,403] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:20,314] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:21,196] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:21,198] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:22,043] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 12:25:22,359] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:22,361] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:22,900] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:23,098] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:23,100] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:24,173] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:24,596] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:24,598] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:25:25,104] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:25:25,277] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:25:25,279] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:26:56,329] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 12:26:56,338] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 12:26:58,486] INFO - scvi.dataset.dataset | extracting highly variable genes using seurat_v3 flavor\n",
      "[2020-11-02 12:27:24,091] INFO - scvi.dataset.dataset | Downsampling from 27569 to 5000 genes\n",
      "[2020-11-02 12:27:32,143] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:27:32,715] INFO - scvi.dataset.dataset | Filtering non-expressing cells.\n",
      "[2020-11-02 12:27:37,272] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 12:27:37,864] INFO - scvi.dataset.dataset | Downsampled from 165827 to 165827 cells\n"
     ]
    }
   ],
   "source": [
    "# prepare data for scvi, subset out doublets and populate scvi objects \n",
    "sp = lpy.scvi_prepare(adata, \"sample_names\", adata.obs[obskey_filteredcells] == False, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mRun scvi\u001b[0m\u001b[34m\n",
      "def runSCVI(dataset, nbstep = 500, n_latent = 64, doinspect= False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mRun scvi\\033[0m\\033[34m\"); print(inspect.getsource(runSCVI));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    vae = VAE(dataset.nb_genes, n_batch= dataset.n_batches, n_labels= dataset.n_labels, n_latent = n_latent)\n",
      "    trainer = UnsupervisedTrainer(vae, dataset, train_size=0.9, frequency=5, use_cuda=True)\n",
      "    trainer.train(n_epochs=nbstep)\n",
      "    full = trainer.create_posterior(trainer.model, dataset, indices=np.arange(len(dataset)))\n",
      "    return(full.sequential().get_latent()[0])\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "[2020-11-02 12:27:54,096] INFO - scvi.inference.inference | KL warmup for 400 epochs\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1c8e5ebc3ba14f2ba56139ded57a48e7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='training', max=500.0, style=ProgressStyle(description_wid"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# run scvi\n",
    "latent = lpy.runSCVI(sp[\"dataset\"], doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mCompute Clusters and Reduces representations\u001b[0m\u001b[34m\n",
      "def insertLatent(adata, latent , latent_key= \"latent\", umap_key= \"X_umap\", tsne_key = \"X_tsne\", leiden_key = \"leiden\", rename_cluster_key= None,cellfilter = None, cellnames =None, leiden_resolution=1.0,doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mCompute Clusters and Reduces representations\\033[0m\\033[34m\"); print(inspect.getsource(insertLatent));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "\n",
      "\n",
      "    if cellnames is None:\n",
      "        #order of full must match\n",
      "        assert latent.shape[0] == len(adata.obs_names),  \"cell names need for be provided if size of latent mismatches adata\"\n",
      "        map = range(len(adata.obs_names))\n",
      "        if latent_key is not None:\n",
      "            adata.obsm[latent_key] = latent\n",
      "    else:\n",
      "        print(\"defining permutation\")\n",
      "        dalist = list(adata.obs_names)\n",
      "        tmap = {}\n",
      "        for i in range(len(adata.obs_names)):\n",
      "            tmap.update( {adata.obs_names[i] : i})\n",
      "# for i in range(len(adata.obs_names))}\n",
      "        map = [tmap[s] for s in cellnames]\n",
      "        if latent_key is not None:\n",
      "            print(\"Inserting Latent coords\")\n",
      "            adata.obsm[latent_key] = np.zeros( (len(adata.obs_names), latent.shape[1]) )\n",
      "            for i in range(len(map)):\n",
      "                adata.obsm[latent_key][map[i],:] = latent[i,:]\n",
      "    \n",
      "    if umap_key is not None:\n",
      "        import umap.umap_ as umap\n",
      "        print(\"computing UMAP\")\n",
      "        tumap = umap.UMAP(spread=2).fit_transform(latent)\n",
      "        adata.obsm[umap_key] = np.zeros( ( len(adata.obs_names), 2 ) )\n",
      "\n",
      "        print(\"Inserting Umap coords\")\n",
      "        for i in range(len(map)):\n",
      "            adata.obsm[umap_key][map[i],:] = tumap[i,:]\n",
      "\n",
      "    if tsne_key is not None:\n",
      "        import scanpy as sc\n",
      "        print(\"computing Tsne\")\n",
      "        adata_latent = sc.AnnData(latent)\n",
      "\n",
      "        sc.tl.tsne(adata_latent, use_rep='X', n_jobs=8)\n",
      "        adata.obsm[tsne_key] = np.zeros( ( len(adata.obs_names), 2 ) )\n",
      "\n",
      "        print(\"Inserting Tsne coords\")\n",
      "        for i in range(len(map)):\n",
      "            adata.obsm[tsne_key][map[i],:] = adata_latent.obsm[\"X_tsne\"][i,:]\n",
      "\n",
      "    if leiden_key is not None:\n",
      "        import scanpy as sc\n",
      "        adata_latent = sc.AnnData(latent)\n",
      "        print(\"Finding clusters\")\n",
      "        sc.pp.neighbors(adata_latent, use_rep='X', n_neighbors=30, metric='minkowski')\n",
      "        sc.tl.leiden(adata_latent, resolution=leiden_resolution)\n",
      "        adata.obs[leiden_key] = \"filtered\"\n",
      "        #if rename_cluster_key is not None:\n",
      "        #    print(\"Renaming to do...\")\n",
      "        #    for i\n",
      "        #else\n",
      "        #    ctnames = range(200)\n",
      "        \n",
      "        print(\"Inserting Cluster Id\")\n",
      "        for i in range(len(map)):\n",
      "            adata.obs[leiden_key][map[i]] = adata_latent.obs[\"leiden\"][i]\n",
      "    return adata\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "defining permutation\n",
      "Inserting Latent coords\n",
      "computing UMAP\n",
      "Inserting Umap coords\n",
      "Finding clusters\n",
      "Inserting Cluster Id\n"
     ]
    }
   ],
   "source": [
    "# store latent variable back in th object, and produce UMAP coordinates and leiden clusters \n",
    "adata = lpy.insertLatent(adata,latent, \"scvi_sampl_cc\", \"X_umap_scvi_sampl_cc\", None, \"leiden_scvi_sampl_cc\", cellnames = sp[\"names\"], doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mPrepare Data for Scvi/TotalVi\u001b[0m\u001b[34m\n",
      "def scvi_prepare(anndatapath, field, cellfilter = None, nbgenes = 5000, genes_to_filter= None, use_ccfilter_prefix=None, citeseqkey = \"protein_expression\", use_raw_slot_instead =None, min_cell_threshold= 0, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mPrepare Data for Scvi/TotalVi\\033[0m\\033[34m\"); print(inspect.getsource(scvi_prepare));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    if use_ccfilter_prefix is not None :\n",
      "        genes_to_filter = [use_ccfilter_prefix + x for x in [\"HMGB2\",\"CDK1\",\"NUSAP1\",\"UBE2C\",\"BIRC5\",\"TPX2\",\"TOP2A\",\"NDC80\",\"CKS2\",\"NUF2\",\"CKS1B\",\"MKI67\",\"TMPO\",\"CENPF\",\"TACC3\",\"FAM64A\",\"SMC4\",\"CCNB2\",\"CKAP2L\",\"CKAP2\",\"AURKB\",\"BUB1\",\"KIF11\",\"ANP32E\",\"TUBB4B\",\"GTSE1\",\"KIF20B\",\"HJURP\",\"CDCA3\",\"HN1\",\"CDC20\",\"TTK\",\"CDC25C\",\"KIF2C\",\"RANGAP1\",\"NCAPD2\",\"DLGAP5\",\"CDCA2\",\"CDCA8\",\"ECT2\",\"KIF23\",\"HMMR\",\"AURKA\",\"PSRC1\",\"ANLN\",\"LBR\",\"CKAP5\",\"CENPE\",\"CTCF\",\"NEK2\",\"G2E3\",\"GAS2L3\",\"CBX5\",\"CENPA\",\"MCM5\",\"PCNA\",\"TYMS\",\"FEN1\",\"MCM2\",\"MCM4\",\"RRM1\",\"UNG\",\"GINS2\",\"MCM6\",\"CDCA7\",\"DTL\",\"PRIM1\",\"UHRF1\",\"MLF1IP\",\"HELLS\",\"RFC2\",\"RPA2\",\"NASP\",\"RAD51AP1\",\"GMNN\",\"WDR76\",\"SLBP\",\"CCNE2\",\"UBR7\",\"POLD3\",\"MSH2\",\"ATAD2\",\"RAD51\",\"RRM2\",\"CDC45\",\"CDC6\",\"EXO1\",\"TIPIN\",\"DSCC1\",\"BLM\",\"CASP8AP2\",\"USP1\",\"CLSPN\",\"POLA1\",\"CHAF1B\",\"BRIP1\",\"E2F8\"]]\n",
      "    if (isinstance(anndatapath, str)): cite = anndata.read_h5ad(anndatapath)\n",
      "    else: cite = anndatapath\n",
      "    if use_raw_slot_instead is None: # default behavior, check for the existance of the raw slot\n",
      "        use_raw_slot_instead = cite.raw is not None\n",
      "    \n",
      "    split = splitAnnData(cite, field, entryfilter= cellfilter, getnames = True, use_raw_slot_instead= use_raw_slot_instead,min_cell_threshold=min_cell_threshold)\n",
      "    if genes_to_filter is not None:\n",
      "        for i in range(len(split[\"datalist\"])):\n",
      "            split[\"datalist\"][i] = split[\"datalist\"][i][:, [g not in genes_to_filter for g in split[\"datalist\"][i].var_names]]\n",
      "    \n",
      "    \n",
      "\n",
      "    dataset = GeneExpressionDataset()\n",
      "    if citeseqkey in cite.obsm.keys():\n",
      "        if citeseqkey is not \"protein_expression\":\n",
      "            for i in range(len(split[\"datalist\"])):\n",
      "                split[\"datalist\"][i].obsm[\"protein_expression\"] = split[\"datalist\"][i].obsm[citeseqkey]\n",
      "        dataset.populate_from_datasets([AnnDatasetFromAnnData(ad=s,cell_measurements_col_mappings={\"protein_expression\":\"protein_names\"})   for s in split[\"datalist\"] ])\n",
      "    else:\n",
      "        dataset.populate_from_datasets([AnnDatasetFromAnnData(ad=s)   for s in split[\"datalist\"] ])\n",
      "\n",
      "    if nbgenes != 0:\n",
      "        dataset.subsample_genes(nbgenes)\n",
      "    return{\"dataset\":dataset , \"names\" : split[\"ordered\"]}\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "[2020-11-02 14:22:24,604] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:22:25,538] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:25,541] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:27,771] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:22:28,117] INFO - scvi.dataset.dataset | Downsampled from 7223 to 7223 cells\n",
      "[2020-11-02 14:22:28,148] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:22:29,173] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:29,175] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:31,563] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:22:31,949] INFO - scvi.dataset.dataset | Downsampled from 8056 to 8056 cells\n",
      "[2020-11-02 14:22:31,983] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:22:32,833] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:32,835] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:34,769] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:22:35,086] INFO - scvi.dataset.dataset | Downsampled from 6627 to 6627 cells\n",
      "[2020-11-02 14:22:35,114] INFO - scvi.dataset.dataset | Merging datasets. Input objects are modified in place.\n",
      "[2020-11-02 14:22:35,116] INFO - scvi.dataset.dataset | Gene names and cell measurement names are assumed to have a non-null intersection between datasets.\n",
      "[2020-11-02 14:22:35,176] INFO - scvi.dataset.dataset | Keeping 27569 genes\n",
      "[2020-11-02 14:22:37,427] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:22:38,307] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:38,309] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:40,829] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:22:41,828] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:41,831] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:43,882] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:22:44,679] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:44,681] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:48,937] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:22:48,939] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:22:50,004] INFO - scvi.dataset.dataset | extracting highly variable genes using seurat_v3 flavor\n",
      "[2020-11-02 14:23:03,040] INFO - scvi.dataset.dataset | Downsampling from 27569 to 5000 genes\n",
      "[2020-11-02 14:23:04,770] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:23:05,069] INFO - scvi.dataset.dataset | Filtering non-expressing cells.\n",
      "[2020-11-02 14:23:06,551] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:23:06,736] INFO - scvi.dataset.dataset | Downsampled from 21906 to 21906 cells\n"
     ]
    }
   ],
   "source": [
    "# prepare data for scvi, subset out doublets and populate scvi objects\n",
    "# zoom-in on anaylisis with \"bulk\" organoids only\n",
    "sp = lpy.scvi_prepare(adata, \"demultiplexed\", (adata.obs[obskey_filteredcells] == False)& (adata.obs[\"demultiplexed\"] != \"B19\") & (adata.obs[\"Lane\"] == 12), doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mRun scvi\u001b[0m\u001b[34m\n",
      "def runSCVI(dataset, nbstep = 500, n_latent = 64, doinspect= False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mRun scvi\\033[0m\\033[34m\"); print(inspect.getsource(runSCVI));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    vae = VAE(dataset.nb_genes, n_batch= dataset.n_batches, n_labels= dataset.n_labels, n_latent = n_latent)\n",
      "    trainer = UnsupervisedTrainer(vae, dataset, train_size=0.9, frequency=5, use_cuda=True)\n",
      "    trainer.train(n_epochs=nbstep)\n",
      "    full = trainer.create_posterior(trainer.model, dataset, indices=np.arange(len(dataset)))\n",
      "    return(full.sequential().get_latent()[0])\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "[2020-11-02 14:23:09,986] INFO - scvi.inference.inference | KL warmup for 400 epochs\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fd9f7bf3ac9749d6bee783deb3788f0b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='training', max=500.0, style=ProgressStyle(description_wid"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# run scvi\n",
    "latent = lpy.runSCVI(sp[\"dataset\"], doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mPrepare Data for Scvi/TotalVi\u001b[0m\u001b[34m\n",
      "def scvi_prepare(anndatapath, field, cellfilter = None, nbgenes = 5000, genes_to_filter= None, use_ccfilter_prefix=None, citeseqkey = \"protein_expression\", use_raw_slot_instead =None, min_cell_threshold= 0, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mPrepare Data for Scvi/TotalVi\\033[0m\\033[34m\"); print(inspect.getsource(scvi_prepare));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    if use_ccfilter_prefix is not None :\n",
      "        genes_to_filter = [use_ccfilter_prefix + x for x in [\"HMGB2\",\"CDK1\",\"NUSAP1\",\"UBE2C\",\"BIRC5\",\"TPX2\",\"TOP2A\",\"NDC80\",\"CKS2\",\"NUF2\",\"CKS1B\",\"MKI67\",\"TMPO\",\"CENPF\",\"TACC3\",\"FAM64A\",\"SMC4\",\"CCNB2\",\"CKAP2L\",\"CKAP2\",\"AURKB\",\"BUB1\",\"KIF11\",\"ANP32E\",\"TUBB4B\",\"GTSE1\",\"KIF20B\",\"HJURP\",\"CDCA3\",\"HN1\",\"CDC20\",\"TTK\",\"CDC25C\",\"KIF2C\",\"RANGAP1\",\"NCAPD2\",\"DLGAP5\",\"CDCA2\",\"CDCA8\",\"ECT2\",\"KIF23\",\"HMMR\",\"AURKA\",\"PSRC1\",\"ANLN\",\"LBR\",\"CKAP5\",\"CENPE\",\"CTCF\",\"NEK2\",\"G2E3\",\"GAS2L3\",\"CBX5\",\"CENPA\",\"MCM5\",\"PCNA\",\"TYMS\",\"FEN1\",\"MCM2\",\"MCM4\",\"RRM1\",\"UNG\",\"GINS2\",\"MCM6\",\"CDCA7\",\"DTL\",\"PRIM1\",\"UHRF1\",\"MLF1IP\",\"HELLS\",\"RFC2\",\"RPA2\",\"NASP\",\"RAD51AP1\",\"GMNN\",\"WDR76\",\"SLBP\",\"CCNE2\",\"UBR7\",\"POLD3\",\"MSH2\",\"ATAD2\",\"RAD51\",\"RRM2\",\"CDC45\",\"CDC6\",\"EXO1\",\"TIPIN\",\"DSCC1\",\"BLM\",\"CASP8AP2\",\"USP1\",\"CLSPN\",\"POLA1\",\"CHAF1B\",\"BRIP1\",\"E2F8\"]]\n",
      "    if (isinstance(anndatapath, str)): cite = anndata.read_h5ad(anndatapath)\n",
      "    else: cite = anndatapath\n",
      "    if use_raw_slot_instead is None: # default behavior, check for the existance of the raw slot\n",
      "        use_raw_slot_instead = cite.raw is not None\n",
      "    \n",
      "    split = splitAnnData(cite, field, entryfilter= cellfilter, getnames = True, use_raw_slot_instead= use_raw_slot_instead,min_cell_threshold=min_cell_threshold)\n",
      "    if genes_to_filter is not None:\n",
      "        for i in range(len(split[\"datalist\"])):\n",
      "            split[\"datalist\"][i] = split[\"datalist\"][i][:, [g not in genes_to_filter for g in split[\"datalist\"][i].var_names]]\n",
      "    \n",
      "    \n",
      "\n",
      "    dataset = GeneExpressionDataset()\n",
      "    if citeseqkey in cite.obsm.keys():\n",
      "        if citeseqkey is not \"protein_expression\":\n",
      "            for i in range(len(split[\"datalist\"])):\n",
      "                split[\"datalist\"][i].obsm[\"protein_expression\"] = split[\"datalist\"][i].obsm[citeseqkey]\n",
      "        dataset.populate_from_datasets([AnnDatasetFromAnnData(ad=s,cell_measurements_col_mappings={\"protein_expression\":\"protein_names\"})   for s in split[\"datalist\"] ])\n",
      "    else:\n",
      "        dataset.populate_from_datasets([AnnDatasetFromAnnData(ad=s)   for s in split[\"datalist\"] ])\n",
      "\n",
      "    if nbgenes != 0:\n",
      "        dataset.subsample_genes(nbgenes)\n",
      "    return{\"dataset\":dataset , \"names\" : split[\"ordered\"]}\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "[2020-11-02 14:34:08,908] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:09,157] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:09,159] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:09,738] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:09,819] INFO - scvi.dataset.dataset | Downsampled from 1860 to 1860 cells\n",
      "[2020-11-02 14:34:09,829] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:10,113] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:10,115] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:10,800] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:10,896] INFO - scvi.dataset.dataset | Downsampled from 2185 to 2185 cells\n",
      "[2020-11-02 14:34:11,109] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:11,111] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:11,564] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:11,640] INFO - scvi.dataset.dataset | Downsampled from 9210 to 9210 cells\n",
      "[2020-11-02 14:34:11,644] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:12,146] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:12,148] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:13,333] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:13,507] INFO - scvi.dataset.dataset | Downsampled from 3941 to 3941 cells\n",
      "[2020-11-02 14:34:13,524] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:13,772] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:13,774] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:14,367] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:14,448] INFO - scvi.dataset.dataset | Downsampled from 1836 to 1836 cells\n",
      "[2020-11-02 14:34:14,458] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:14,695] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:14,697] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:15,278] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:15,357] INFO - scvi.dataset.dataset | Downsampled from 1788 to 1788 cells\n",
      "[2020-11-02 14:34:15,367] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:16,097] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:16,099] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:17,788] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:18,051] INFO - scvi.dataset.dataset | Downsampled from 5944 to 5944 cells\n",
      "[2020-11-02 14:34:18,075] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:18,475] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:18,477] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:19,424] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:19,562] INFO - scvi.dataset.dataset | Downsampled from 3146 to 3146 cells\n",
      "[2020-11-02 14:34:19,577] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:19,808] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:19,810] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:20,340] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:20,416] INFO - scvi.dataset.dataset | Downsampled from 1745 to 1745 cells\n",
      "[2020-11-02 14:34:20,426] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:20,781] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:20,783] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:21,595] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:21,717] INFO - scvi.dataset.dataset | Downsampled from 2773 to 2773 cells\n",
      "[2020-11-02 14:34:21,730] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:22,028] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:22,030] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:22,709] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:22,810] INFO - scvi.dataset.dataset | Downsampled from 2303 to 2303 cells\n",
      "[2020-11-02 14:34:22,821] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:23,148] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:23,150] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:23,904] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 14:34:24,013] INFO - scvi.dataset.dataset | Downsampled from 2486 to 2486 cells\n",
      "[2020-11-02 14:34:24,025] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:24,404] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:24,406] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:25,254] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:25,385] INFO - scvi.dataset.dataset | Downsampled from 2988 to 2988 cells\n",
      "[2020-11-02 14:34:25,399] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:25,849] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:25,851] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:26,946] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:27,103] INFO - scvi.dataset.dataset | Downsampled from 3563 to 3563 cells\n",
      "[2020-11-02 14:34:27,119] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:27,719] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:27,721] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:29,076] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:29,287] INFO - scvi.dataset.dataset | Downsampled from 4815 to 4815 cells\n",
      "[2020-11-02 14:34:29,307] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:29,759] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:29,761] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:30,761] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:30,917] INFO - scvi.dataset.dataset | Downsampled from 3547 to 3547 cells\n",
      "[2020-11-02 14:34:30,933] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:31,218] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:31,220] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:31,881] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:31,977] INFO - scvi.dataset.dataset | Downsampled from 2181 to 2181 cells\n",
      "[2020-11-02 14:34:31,988] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:32,280] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:32,282] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:32,970] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:33,068] INFO - scvi.dataset.dataset | Downsampled from 2223 to 2223 cells\n",
      "[2020-11-02 14:34:33,079] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:33,436] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:33,438] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:34,250] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:34,369] INFO - scvi.dataset.dataset | Downsampled from 2761 to 2761 cells\n",
      "[2020-11-02 14:34:34,382] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:34,640] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:34,642] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:35,250] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:35,336] INFO - scvi.dataset.dataset | Downsampled from 1968 to 1968 cells\n",
      "[2020-11-02 14:34:35,346] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:35,692] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:35,694] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:36,496] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:36,613] INFO - scvi.dataset.dataset | Downsampled from 2652 to 2652 cells\n",
      "[2020-11-02 14:34:36,626] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:36,818] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:36,820] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:37,276] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:37,339] INFO - scvi.dataset.dataset | Downsampled from 1399 to 1399 cells\n",
      "[2020-11-02 14:34:37,348] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:37,602] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:37,604] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:38,179] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:38,254] INFO - scvi.dataset.dataset | Downsampled from 1676 to 1676 cells\n",
      "[2020-11-02 14:34:38,264] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:38,765] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:38,767] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:39,848] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:40,021] INFO - scvi.dataset.dataset | Downsampled from 3899 to 3899 cells\n",
      "[2020-11-02 14:34:40,039] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:40,290] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:40,292] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:40,854] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:40,933] INFO - scvi.dataset.dataset | Downsampled from 1790 to 1790 cells\n",
      "[2020-11-02 14:34:40,943] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:41,357] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:41,359] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:42,345] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:42,487] INFO - scvi.dataset.dataset | Downsampled from 3211 to 3211 cells\n",
      "[2020-11-02 14:34:42,502] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:43,284] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:43,286] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:45,055] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:45,322] INFO - scvi.dataset.dataset | Downsampled from 6137 to 6137 cells\n",
      "[2020-11-02 14:34:45,347] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:46,195] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:46,197] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:48,113] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:48,414] INFO - scvi.dataset.dataset | Downsampled from 6849 to 6849 cells\n",
      "[2020-11-02 14:34:48,442] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:48,792] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:48,794] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:49,589] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 14:34:49,704] INFO - scvi.dataset.dataset | Downsampled from 2632 to 2632 cells\n",
      "[2020-11-02 14:34:49,716] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:50,139] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:50,141] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:51,104] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:51,249] INFO - scvi.dataset.dataset | Downsampled from 3292 to 3292 cells\n",
      "[2020-11-02 14:34:51,264] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:51,652] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:51,654] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:52,554] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:52,686] INFO - scvi.dataset.dataset | Downsampled from 2983 to 2983 cells\n",
      "[2020-11-02 14:34:52,700] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:53,385] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:53,387] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:55,052] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:55,289] INFO - scvi.dataset.dataset | Downsampled from 5415 to 5415 cells\n",
      "[2020-11-02 14:34:55,312] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:55,768] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:55,770] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:56,873] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:57,031] INFO - scvi.dataset.dataset | Downsampled from 3599 to 3599 cells\n",
      "[2020-11-02 14:34:57,047] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:57,385] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:57,386] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:34:58,177] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:34:58,292] INFO - scvi.dataset.dataset | Downsampled from 2600 to 2600 cells\n",
      "[2020-11-02 14:34:58,304] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:34:58,840] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:34:58,842] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:00,100] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:00,283] INFO - scvi.dataset.dataset | Downsampled from 4183 to 4183 cells\n",
      "[2020-11-02 14:35:00,301] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:00,685] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:00,687] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:01,558] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:01,690] INFO - scvi.dataset.dataset | Downsampled from 3004 to 3004 cells\n",
      "[2020-11-02 14:35:01,704] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:02,223] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:02,225] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:03,439] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:03,623] INFO - scvi.dataset.dataset | Downsampled from 4086 to 4086 cells\n",
      "[2020-11-02 14:35:03,641] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:03,891] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:03,893] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:04,447] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:04,526] INFO - scvi.dataset.dataset | Downsampled from 1785 to 1785 cells\n",
      "[2020-11-02 14:35:04,535] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:05,118] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:05,121] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:06,398] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:06,604] INFO - scvi.dataset.dataset | Downsampled from 4675 to 4675 cells\n",
      "[2020-11-02 14:35:06,624] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:06,864] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:06,866] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:07,431] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:07,510] INFO - scvi.dataset.dataset | Downsampled from 1801 to 1801 cells\n",
      "[2020-11-02 14:35:07,520] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:07,860] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:07,862] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:08,643] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:08,756] INFO - scvi.dataset.dataset | Downsampled from 2586 to 2586 cells\n",
      "[2020-11-02 14:35:08,770] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:09,500] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:09,501] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:11,166] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:11,423] INFO - scvi.dataset.dataset | Downsampled from 5828 to 5828 cells\n",
      "[2020-11-02 14:35:11,448] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:11,817] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:11,818] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:12,633] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:12,749] INFO - scvi.dataset.dataset | Downsampled from 2624 to 2624 cells\n",
      "[2020-11-02 14:35:12,762] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:13,205] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:13,207] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:14,242] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:14,395] INFO - scvi.dataset.dataset | Downsampled from 3496 to 3496 cells\n",
      "[2020-11-02 14:35:14,411] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:14,701] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:14,703] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:15,390] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:15,487] INFO - scvi.dataset.dataset | Downsampled from 2211 to 2211 cells\n",
      "[2020-11-02 14:35:15,498] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:15,602] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:15,604] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:15,803] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 14:35:15,831] INFO - scvi.dataset.dataset | Downsampled from 629 to 629 cells\n",
      "[2020-11-02 14:35:15,836] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:16,117] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:16,119] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:16,742] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:16,835] INFO - scvi.dataset.dataset | Downsampled from 2122 to 2122 cells\n",
      "[2020-11-02 14:35:16,846] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:17,610] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:17,612] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:19,411] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:19,685] INFO - scvi.dataset.dataset | Downsampled from 6208 to 6208 cells\n",
      "[2020-11-02 14:35:19,711] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:20,050] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:20,052] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:20,834] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:20,948] INFO - scvi.dataset.dataset | Downsampled from 2590 to 2590 cells\n",
      "[2020-11-02 14:35:20,960] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:21,183] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:21,185] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:21,685] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:21,756] INFO - scvi.dataset.dataset | Downsampled from 1629 to 1629 cells\n",
      "[2020-11-02 14:35:21,766] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:22,216] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:22,218] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:23,208] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:23,363] INFO - scvi.dataset.dataset | Downsampled from 3526 to 3526 cells\n",
      "[2020-11-02 14:35:23,379] INFO - scvi.dataset.anndataset | Dense size under 1Gb, casting to dense format (np.ndarray).\n",
      "[2020-11-02 14:35:23,579] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:23,581] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:24,062] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:24,126] INFO - scvi.dataset.dataset | Downsampled from 1447 to 1447 cells\n",
      "[2020-11-02 14:35:24,134] INFO - scvi.dataset.dataset | Merging datasets. Input objects are modified in place.\n",
      "[2020-11-02 14:35:24,135] INFO - scvi.dataset.dataset | Gene names and cell measurement names are assumed to have a non-null intersection between datasets.\n",
      "[2020-11-02 14:35:25,209] INFO - scvi.dataset.dataset | Keeping 27569 genes\n",
      "[2020-11-02 14:35:25,835] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:26,059] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:26,061] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:26,779] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:27,037] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:27,039] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:27,587] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:27,669] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:27,671] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:28,922] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:29,414] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:29,416] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:30,036] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:30,257] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:30,259] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:30,865] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:31,078] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:31,080] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:32,871] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:33,575] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:33,577] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:34,576] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:34,966] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:34,968] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:35,532] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:35,742] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:35,744] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:36,603] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:36,924] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:36,926] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:37,648] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:37,918] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:37,920] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:38,720] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:39,032] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:39,034] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:39,978] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:40,335] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:40,337] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:41,484] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:41,898] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:41,900] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:43,343] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:43,905] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:43,907] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:44,976] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:45,386] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:45,388] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:46,088] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:46,344] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:46,346] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:47,067] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:47,328] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:47,330] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 14:35:48,198] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:48,519] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:48,521] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:49,162] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:49,402] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:49,404] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:50,254] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:50,569] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:50,571] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:51,037] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:51,202] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:51,205] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:51,762] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:51,969] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:51,971] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:53,135] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:53,590] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:53,592] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:54,186] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:54,409] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:54,410] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:55,444] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:55,822] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:55,824] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:35:57,691] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:35:58,405] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:35:58,407] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:00,444] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:01,244] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:01,246] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:02,079] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:02,396] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:02,398] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:03,421] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:03,821] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:03,822] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:04,772] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:05,124] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:05,126] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:06,856] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:07,480] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:07,482] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:08,640] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:09,061] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:09,063] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:09,893] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:10,207] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:10,209] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:11,515] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:12,004] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:12,006] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:12,936] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:13,301] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:13,303] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:14,567] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:15,055] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:15,057] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:15,648] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:15,859] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:15,861] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:17,236] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:17,779] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:17,781] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:18,382] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:18,594] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:18,596] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:19,461] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:19,777] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:19,779] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:21,555] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:22,264] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:22,266] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:23,109] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:23,480] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:23,482] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:24,576] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:24,994] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:24,996] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:25,722] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:25,979] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:25,981] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:26,203] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:26,283] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:26,284] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:26,952] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:27,215] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:27,217] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:29,098] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:29,968] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:29,970] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:30,798] INFO - scvi.dataset.dataset | Computing the library size for the new data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-11-02 14:36:31,109] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:31,111] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:31,647] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:31,837] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:31,839] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:32,902] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:33,330] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:33,331] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:36:33,835] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:36:34,008] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:36:34,010] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:38:05,651] INFO - scvi.dataset.dataset | Remapping labels to [0,N]\n",
      "[2020-11-02 14:38:05,658] INFO - scvi.dataset.dataset | Remapping batch_indices to [0,N]\n",
      "[2020-11-02 14:38:07,787] INFO - scvi.dataset.dataset | extracting highly variable genes using seurat_v3 flavor\n",
      "[2020-11-02 14:38:30,703] INFO - scvi.dataset.dataset | Downsampling from 27569 to 5000 genes\n",
      "[2020-11-02 14:38:38,639] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:38:39,188] INFO - scvi.dataset.dataset | Filtering non-expressing cells.\n",
      "[2020-11-02 14:38:43,704] INFO - scvi.dataset.dataset | Computing the library size for the new data\n",
      "[2020-11-02 14:38:44,235] INFO - scvi.dataset.dataset | Downsampled from 165827 to 165827 cells\n"
     ]
    }
   ],
   "source": [
    "# prepare data for scvi, subset out doublets and populate scvi objects \n",
    "sp = lpy.scvi_prepare(adata, \"sample_names\", adata.obs[obskey_filteredcells] == False, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'Hormones' as categorical\n",
      "... storing 'Inhibitor' as categorical\n",
      "... storing 'leiden_scvi_sampl_cc' as categorical\n"
     ]
    }
   ],
   "source": [
    "# save object\n",
    "adata.write_h5ad(\"N1-integrated_organoids.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
