{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.6.0 anndata==0.7.4 umap==0.4.6 numpy==1.19.4 scipy==1.5.3 pandas==1.1.4 scikit-learn==0.22.2 statsmodels==0.11.1 python-igraph==0.8.0 louvain==0.6.1 leidenalg==0.8.2\n"
     ]
    }
   ],
   "source": [
    "# Integration of biopsies and organ donors from 2 studiesar\n",
    "import anndata, numpy as np, pandas as pd, imp, lpy, scanpy as sc\n",
    "sc.logging.print_header()\n",
    "%load_ext rpy2.ipython\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set some constants\n",
    "sampledir = \"/lustre/scratch117/cellgen/team292/lh20/organoids/\"\n",
    "demul = pd.read_csv(\"demultiplex.csv\",delimiter= '\\t')\n",
    "\n",
    "samplelist = demul[\"sampleID\"]\n",
    "souporcell_genodico = {demul[\"sampleID\"][i] : [demul[\"genotype0\"][i], demul[\"genotype1\"][i], demul[\"genotype2\"][i], demul[\"genotype3\"][i]] for i in range(demul.shape[0])}\n",
    "folderlist = [sampledir +x + \"/\" for x in samplelist]\n",
    "soupxfolderlist = [sampledir + x + \"/\" for x in samplelist ] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mCreate AnnData Object\u001b[0m\u001b[34m\n",
      "def createAnnData(folderlist, prefix, souporcell_folderlist = None, souporcell_genodico = None, autoinclude=[\"percent_mito\", \"log2p1_count\", \"n_genes\"], min_cell_per_gene_allowed=3, min_gene_per_cell_allowed=500, sample_obskey = \"sample_names\",doqcplots=False, doinspect=False, mitogeneprefix=\"MT-\", do_log2_normalize=True):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mCreate AnnData Object\\033[0m\\033[34m\"); print(inspect.getsource(createAnnData));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    adatas = []\n",
      "    def pickname(x,y):\n",
      "        if (y == \"singlet\"): return(\"_genotype_\" +str(x))\n",
      "        else: return(\"_doublet\")\n",
      "    def pickname2(x,y,z):\n",
      "        if (y == \"singlet\"): return(z[int(x)])\n",
      "        else: return(\"doublet\")\n",
      "    for i in range(len(folderlist)):\n",
      "        print(\"Processing \" + prefix[i])\n",
      "        adatas.append(sc.read_10x_mtx(folderlist[i]))\n",
      "        if souporcell_folderlist is not None:\n",
      "            try:\n",
      "                res = pd.read_csv(souporcell_folderlist[i] + \"clusters.tsv\",sep='\\t')\n",
      "                if (res.shape[0] != len(adatas[i].obs_names)): raise ValueError( prefix[i] + \" data and soupx barcodes mismatches!\")\n",
      "                adatas[i].obs[\"souporcell\"] = res.status.to_list()\n",
      "                if (souporcell_genodico is None) or (prefix[i] not in souporcell_genodico.keys()):\n",
      "                    adatas[i].obs[\"demultiplexed\"] = [prefix[i] + pickname(x,y)   for x,y in zip( res.assignment.to_list() , res.status.to_list()) ]\n",
      "                else:\n",
      "                    adatas[i].obs[\"demultiplexed\"] = [pickname2(x,y, souporcell_genodico[prefix[i]] )   for x,y in zip( res.assignment.to_list() , res.status.to_list()) ]\n",
      "            except:\n",
      "                print(\"No valid genotyping data for \"+ str(len(adatas[i].obs_names)) + \" cells within \" + prefix[i] + \"! Setting everything to singlets\")\n",
      "                adatas[i].obs[\"souporcell\"] = \"singlet\"\n",
      "                if (souporcell_genodico is None):\n",
      "                    adatas[i].obs[\"demultiplexed\"] = prefix[i] + \"_genotype_0\"\n",
      "                else:\n",
      "                    adatas[i].obs[\"demultiplexed\"] = souporcell_genodico[prefix[i]][0]\n",
      "                    \n",
      "        adatas[i].obs[sample_obskey] = [prefix[i] for hehe in adatas[i].obs_names]\n",
      "        adatas[i].obs_names = [prefix[i] + \"_\" + x for x in adatas[i].obs_names]\n",
      "        if \"n_count\" in autoinclude:\n",
      "            adatas[i].obs['n_count'] = np.sum(adatas[i].X, axis=1).A1\n",
      "        if \"log2p1_count\" in autoinclude:\n",
      "            adatas[i].obs['log2p1_count'] = np.log1p(np.sum(adatas[i].X, axis=1).A1) / math.log(2)\n",
      "        if \"percent_mito\" in autoinclude:\n",
      "            mito_genes = [name for name in adatas[i].var_names if name.startswith(mitogeneprefix)]\n",
      "            adatas[i].obs['percent_mito'] = np.sum(adatas[i][:, mito_genes].X, axis=1).A1 / np.sum(adatas[i].X, axis=1).A1\n",
      "        if \"n_genes\" in autoinclude:\n",
      "            adatas[i].obs['n_genes'] = np.sum(adatas[i].X != 0, axis=1).A1\n",
      "        adatas[i].obs[sample_obskey].value_counts()\n",
      "    adata = mergeAnnData(adatas,index_unique = None)\n",
      "    if min_gene_per_cell_allowed is not None: sc.pp.filter_cells(adata, min_genes=min_gene_per_cell_allowed)\n",
      "    if min_cell_per_gene_allowed is not None: sc.pp.filter_genes(adata, min_cells=min_cell_per_gene_allowed)\n",
      "    \n",
      "    if doqcplots is True:\n",
      "        if \"log2p1_count\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['log2p1_count'], jitter=0.4, groupby=sample_obskey, rotation=90)\n",
      "        if \"n_genes\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['n_genes'], jitter=0.4, groupby=sample_obskey, rotation=90)\n",
      "        if \"percent_mito\" in autoinclude:\n",
      "            sc.pl.violin(adata, ['percent_mito'], jitter=0.4, groupby=sample_obskey, rotation=90)\n",
      "    if do_log2_normalize:\n",
      "        adata.raw = adata.copy()\n",
      "        adata.X = np.log1p(adata.X) / math.log(2)\n",
      "    if \"log2p1_count\" in autoinclude:\n",
      "        adata.obs[\"log2p1_count\"] = pd.to_numeric(adata.obs[\"log2p1_count\"], downcast='float')\n",
      "    if \"percent_mito\" in autoinclude:\n",
      "        adata.obs[\"percent_mito\"] = pd.to_numeric(adata.obs[\"percent_mito\"], downcast='float')\n",
      "    return adata;\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "Processing organoids7090723\n",
      "No valid genotyping data for 4187 cells within organoids7090723! Setting everything to singlets\n",
      "Processing organoids7090724\n",
      "No valid genotyping data for 6220 cells within organoids7090724! Setting everything to singlets\n",
      "Processing organoids7090725\n",
      "No valid genotyping data for 2644 cells within organoids7090725! Setting everything to singlets\n",
      "Processing organoids7090726\n",
      "No valid genotyping data for 2652 cells within organoids7090726! Setting everything to singlets\n",
      "Processing organoids7090727\n",
      "No valid genotyping data for 2804 cells within organoids7090727! Setting everything to singlets\n",
      "Processing organoids7090728\n",
      "No valid genotyping data for 3408 cells within organoids7090728! Setting everything to singlets\n",
      "Processing organoids7090729\n",
      "No valid genotyping data for 3540 cells within organoids7090729! Setting everything to singlets\n",
      "Processing organoids7090730\n",
      "No valid genotyping data for 3657 cells within organoids7090730! Setting everything to singlets\n",
      "Processing organoids7090731\n",
      "No valid genotyping data for 4712 cells within organoids7090731! Setting everything to singlets\n",
      "Processing organoids7090732\n",
      "No valid genotyping data for 4034 cells within organoids7090732! Setting everything to singlets\n",
      "Processing WSSS_END9263313\n",
      "Processing WSSS_END9263314\n",
      "Processing WSSS_END9263315\n",
      "Processing WSSS_END9263316\n",
      "Processing WSSS_END9263317\n",
      "Processing WSSS_END9397520\n",
      "Processing WSSS_END9397521\n",
      "Processing WSSS_END9397522\n",
      "Processing WSSS_END9397523\n",
      "Processing WSSS_END9397524\n",
      "Processing WSSS_END9397525\n",
      "Processing WSSS_END9397526\n",
      "Processing WSSS_END9397527\n",
      "Processing WSSS_END9397528\n",
      "Processing WSSS_END9397529\n",
      "Processing WSSS_END9397530\n",
      "Processing WSSS_END9397531\n",
      "Processing WSSS_END9397532\n",
      "Processing WSSS_END9397533\n",
      "Processing WSSS_END9397534\n",
      "Processing WSSS_END9397535\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'souporcell' as categorical\n",
      "... storing 'demultiplexed' as categorical\n",
      "... storing 'sample_names' as categorical\n"
     ]
    }
   ],
   "source": [
    "#Load data\n",
    "adata=lpy.createAnnData( folderlist, samplelist, souporcell_folderlist = soupxfolderlist, souporcell_genodico=souporcell_genodico, doqcplots=True, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mAdd metadata from sample table to anndata\u001b[0m\u001b[34m\n",
      "def addMetadata(adata, metadata, obs_key, meta_key, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mAdd metadata from sample table to anndata\\033[0m\\033[34m\"); print(inspect.getsource(addMetadata));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    aslist = metadata[meta_key].tolist()\n",
      "    rowmap = {i : aslist.index(i)  for i in aslist}\n",
      "    for val in metadata.columns:\n",
      "        if val != meta_key:\n",
      "            aslist = metadata[val].tolist()\n",
      "            adata.obs[val] = [aslist[rowmap[i]] for i in adata.obs[obs_key] ]\n",
      "    return adata;\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "WSSS_END9397532     19488\n",
       "WSSS_END9397529     16256\n",
       "WSSS_END9397534     15286\n",
       "WSSS_END9397531     12526\n",
       "WSSS_END9397525     11941\n",
       "WSSS_END9397535     10602\n",
       "WSSS_END9263316     10269\n",
       "WSSS_END9397527     10010\n",
       "WSSS_END9397526      9740\n",
       "WSSS_END9397533      9556\n",
       "WSSS_END9397530      8088\n",
       "WSSS_END9397524      7967\n",
       "WSSS_END9397528      7854\n",
       "WSSS_END9263314      7581\n",
       "WSSS_END9397520      6958\n",
       "WSSS_END9263315      6892\n",
       "organoids7090724     6220\n",
       "WSSS_END9397523      6112\n",
       "WSSS_END9263313      5889\n",
       "WSSS_END9397522      5598\n",
       "organoids7090731     4711\n",
       "WSSS_END9397521      4702\n",
       "organoids7090723     4187\n",
       "organoids7090732     4032\n",
       "organoids7090730     3657\n",
       "organoids7090729     3540\n",
       "organoids7090728     3408\n",
       "organoids7090727     2803\n",
       "organoids7090726     2652\n",
       "organoids7090725     2644\n",
       "WSSS_END9263317       699\n",
       "Name: sample_names, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(\"sample_meta.csv\",delimiter= '\\t',)\n",
    "adata = lpy.addMetadata(adata,meta, \"sample_names\",\"sample_names\",doinspect=True)\n",
    "adata.obs[\"sample_names\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mDetect Doublets and defining cells to filter\u001b[0m\u001b[34m\n",
      "def scrub(adata, batch_obsattrib, bonf_threshold = 0.01, add_qc_metrics=False,mito_prefix= \"MT-\", obskey_cellfilter = \"filtered_cells\", add_cell_filter={\"max_percent_mito\": 0.15, \"scrublet_local_pred\": False}, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mDetect Doublets and defining cells to filter\\033[0m\\033[34m\"); print(inspect.getsource(scrub));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "    \n",
      "    import scrublet as scr\n",
      "    import scanpy as sc\n",
      "    print(\"spliting data using attribute \" + batch_obsattrib)\n",
      "    adatas = splitAnnData(adata, batch_obsattrib)\n",
      "\n",
      "    if (add_qc_metrics):\n",
      "        mito_genes = [name for name in adata.var_names if name.startswith(mito_prefix)]\n",
      "        adata.obs['log2p1_RNA_count'] = np.log1p(adata.X.sum(axis=1).A1) / math.log(2)\n",
      "        adata.obs['percent_mito'] = np.sum(adata[:, mito_genes].X, axis=1).A1 / np.sum(adata.X, axis=1).A1\n",
      "\n",
      "    dalist = list(adata.obs_names)\n",
      "    tmap = {}\n",
      "    for i in range(len(adata.obs_names)):\n",
      "        tmap.update( {adata.obs_names[i] : i})\n",
      "    adata.obs[\"scrublet_pred\"] = False\n",
      "    adata.obs[\"scrublet_local_pred\"] = False\n",
      "    adata.obs[\"scrublet_score\"] = 0.0\n",
      "    adata.obs[\"scrublet_cluster_score\"] = 0.0\n",
      "\n",
      "    # Luz's double clustering approach\n",
      "    for i in range(len(adatas)):\n",
      "\n",
      "        print(\"processing \" + str(i) + \"/\" + str(len(adatas)))\n",
      "        curscr = scr.Scrublet(adatas[i].X)\n",
      "        doublet_scores, predicted_doublets = curscr.scrub_doublets()\n",
      "        adatas[i].obs['scrublet_score'] = doublet_scores\n",
      "        adatas[i].obs['scrublet_pred'] = predicted_doublets\n",
      "        #overcluster prep. run turbo basic scanpy pipeline\n",
      "        sc.pp.normalize_per_cell(adatas[i], counts_per_cell_after=1e4)\n",
      "        sc.pp.log1p(adatas[i])\n",
      "        sc.pp.highly_variable_genes(adatas[i], min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
      "        adatas[i] = adatas[i][:, adatas[i].var['highly_variable']]\n",
      "        sc.pp.scale(adatas[i], max_value=10)\n",
      "        sc.tl.pca(adatas[i], svd_solver='arpack')\n",
      "        sc.pp.neighbors(adatas[i])\n",
      "        #overclustering proper - do basic clustering first, then cluster each cluster\n",
      "        sc.tl.leiden(adatas[i])\n",
      "        adatas[i].obs['leiden'] = [str(i) for i in adatas[i].obs['leiden']]\n",
      "        for clus in np.unique(adatas[i].obs['leiden']):\n",
      "            adata_sub = adatas[i][adatas[i].obs['leiden']==clus].copy()\n",
      "            sc.tl.leiden(adata_sub)\n",
      "            adata_sub.obs['leiden'] = [clus+','+i for i in adata_sub.obs['leiden']]\n",
      "            adatas[i].obs.loc[adata_sub.obs_names,'leiden'] = adata_sub.obs['leiden']\n",
      "\n",
      "        #compute the cluster scores - the median of Scrublet scores per overclustered cluster\n",
      "        for clus in np.unique(adatas[i].obs['leiden']):\n",
      "            adatas[i].obs.loc[adatas[i].obs['leiden']==clus, 'scrublet_cluster_score'] = \\\n",
      "                np.median(adatas[i].obs.loc[adatas[i].obs['leiden']==clus, 'scrublet_score'])\n",
      "        #now compute doublet p-values. figure out the median and mad (from above-median values) for the distribution\n",
      "        med = np.median(adatas[i].obs['scrublet_cluster_score'])\n",
      "        mask = adatas[i].obs['scrublet_cluster_score']>med\n",
      "        mad = np.median(adatas[i].obs['scrublet_cluster_score'][mask]-med)\n",
      "        #let's do a one-sided test. the Bertie write-up does not address this but it makes sense\n",
      "        zscores = (adatas[i].obs['scrublet_cluster_score'].values - med) / (1.4826 * mad)\n",
      "        #adatas[i].obs['zscore'] = zscores\n",
      "        pvals = 1-scipy.stats.norm.cdf(zscores)\n",
      "        #adatas[i].obs['bh_pval'] = bh(pvals)\n",
      "        pvals = bonf(pvals)\n",
      "        adatas[i].obs['scrublet_local_pred'] = pvals < bonf_threshold\n",
      "        map = [tmap[s] for s in adatas[i].obs_names]\n",
      "        print(\"annoying values\")\n",
      "        for j in range(len(adatas[i].obs_names)):\n",
      "            adata.obs[\"scrublet_pred\" ][map[j]] = adatas[i].obs[\"scrublet_pred\"][j]\n",
      "            adata.obs[\"scrublet_local_pred\"][map[j]] = adatas[i].obs[\"scrublet_local_pred\"][j]\n",
      "            adata.obs[\"scrublet_cluster_score\"][map[j]] =  adatas[i].obs[\"scrublet_cluster_score\"][j]\n",
      "            adata.obs[\"scrublet_score\"][map[j]] =  pvals[j]\n",
      "\n",
      "    adata.obs[\"scrublet_cluster_score\"] = pd.to_numeric(adata.obs[\"scrublet_cluster_score\"], downcast='float')\n",
      "    adata.obs[\"scrublet_score\"] = pd.to_numeric(adata.obs[\"scrublet_score\"], downcast='float')\n",
      "\n",
      "    if add_cell_filter is not None:\n",
      "        curflt = np.zeros(adata.obs.shape[0], dtype=bool) \n",
      "        for k,v in add_cell_filter.items():\n",
      "            if k == \"max_percent_mito\":\n",
      "                print(str(sum(adata.obs['percent_mito'] > v)) + \" cells filtered by mito threshold\")\n",
      "                curflt |= adata.obs['percent_mito'] > v\n",
      "            else:\n",
      "                print(str(sum(adata.obs[k] != v)) + \" cells filtered by \" + k + \" == \" + str(v) + \" criterion\")\n",
      "                curflt |= (adata.obs[k] != v)\n",
      "        adata.obs[obskey_cellfilter] = curflt\n",
      "    return adata\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n",
      "spliting data using attribute sample_names\n",
      "processing 0/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.1 seconds\n",
      "annoying values\n",
      "processing 1/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.73\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 4.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.9%\n",
      "Elapsed time: 10.9 seconds\n",
      "annoying values\n",
      "processing 2/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 16.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.4%\n",
      "Elapsed time: 4.2 seconds\n",
      "annoying values\n",
      "processing 3/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.51\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 34.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.5%\n",
      "Elapsed time: 6.4 seconds\n",
      "annoying values\n",
      "processing 4/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.9%\n",
      "Elapsed time: 10.9 seconds\n",
      "annoying values\n",
      "processing 5/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 1.0%\n",
      "Estimated detectable doublet fraction = 39.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.5%\n",
      "Elapsed time: 6.0 seconds\n",
      "annoying values\n",
      "processing 6/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.69\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 11.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.3%\n",
      "Elapsed time: 7.6 seconds\n",
      "annoying values\n",
      "processing 7/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 1.0%\n",
      "Estimated detectable doublet fraction = 39.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.5%\n",
      "Elapsed time: 6.0 seconds\n",
      "annoying values\n",
      "processing 8/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 2.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.8%\n",
      "Elapsed time: 13.1 seconds\n",
      "annoying values\n",
      "processing 9/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.72\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 2.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 9.7 seconds\n",
      "annoying values\n",
      "processing 10/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.51\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 40.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.1%\n",
      "Elapsed time: 4.8 seconds\n",
      "annoying values\n",
      "processing 11/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.78\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.9%\n",
      "Elapsed time: 17.5 seconds\n",
      "annoying values\n",
      "processing 12/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.70\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 6.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.6%\n",
      "Elapsed time: 6.8 seconds\n",
      "annoying values\n",
      "processing 13/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 1.0%\n",
      "Estimated detectable doublet fraction = 28.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.5%\n",
      "Elapsed time: 16.0 seconds\n",
      "annoying values\n",
      "processing 14/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.73\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 5.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.9%\n",
      "Elapsed time: 12.6 seconds\n",
      "annoying values\n",
      "processing 15/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.77\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 9.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.4%\n",
      "Elapsed time: 14.7 seconds\n",
      "annoying values\n",
      "processing 16/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.78\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 16.6 seconds\n",
      "annoying values\n",
      "processing 17/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.2%\n",
      "Elapsed time: 11.9 seconds\n",
      "annoying values\n",
      "processing 18/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.78\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 15.7 seconds\n",
      "annoying values\n",
      "processing 19/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.83\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.0%\n",
      "Elapsed time: 31.0 seconds\n",
      "annoying values\n",
      "processing 20/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.80\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.6%\n",
      "Elapsed time: 19.5 seconds\n",
      "annoying values\n",
      "processing 21/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.7%\n",
      "Estimated detectable doublet fraction = 14.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.0%\n",
      "Elapsed time: 1.3 seconds\n",
      "annoying values\n",
      "processing 22/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 9.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.5%\n",
      "Elapsed time: 5.5 seconds\n",
      "annoying values\n",
      "processing 23/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 25.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.6%\n",
      "Elapsed time: 11.2 seconds\n",
      "annoying values\n",
      "processing 24/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.33\n",
      "Detected doublet rate = 9.0%\n",
      "Estimated detectable doublet fraction = 51.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 17.4%\n",
      "Elapsed time: 24.5 seconds\n",
      "annoying values\n",
      "processing 25/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.1 seconds\n",
      "annoying values\n",
      "processing 26/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 7.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.4%\n",
      "Elapsed time: 4.9 seconds\n",
      "annoying values\n",
      "processing 27/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.35\n",
      "Detected doublet rate = 8.9%\n",
      "Estimated detectable doublet fraction = 50.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 17.6%\n",
      "Elapsed time: 25.2 seconds\n",
      "annoying values\n",
      "processing 28/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.70\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 7.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.3%\n",
      "Elapsed time: 8.5 seconds\n",
      "annoying values\n",
      "processing 29/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 2.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 12.9 seconds\n",
      "annoying values\n",
      "processing 30/31\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.80\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.0%\n",
      "Elapsed time: 19.6 seconds\n",
      "annoying values\n",
      "20675 cells filtered by souporcell == singlet criterion\n",
      "16860 cells filtered by mito threshold\n",
      "1183 cells filtered by scrublet_local_pred == False criterion\n"
     ]
    }
   ],
   "source": [
    "obskey_filteredcells = \"filtered_cells\"\n",
    "adata=lpy.scrub(adata, \"sample_names\",obskey_cellfilter= obskey_filteredcells,add_cell_filter={\"souporcell\": \"singlet\" ,\"max_percent_mito\":0.3, \"scrublet_local_pred\": False }, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[35;46;1mAdd Cell Cycle annotation to anndata object\u001b[0m\u001b[34m\n",
      "def addCycleCycleAnnotation(adata, s_genes = None, g2m_genes = None, geneprefix = \"\", use_raw_data= True, doinspect=False):\n",
      "    if doinspect is True: print(\"\\033[35;46;1mAdd Cell Cycle annotation to anndata object\\033[0m\\033[34m\"); print(inspect.getsource(addCycleCycleAnnotation));print(\"\\033[31;43;1mExecution:\\033[0m\")\n",
      "\n",
      "    # uses Seurat Cell Cycles default genes by default\n",
      "    if s_genes is None: # \"MLF1IP\"\n",
      "        s_genes = [\"MCM5\",\"PCNA\",\"TYMS\",\"FEN1\",\"MCM2\",\"MCM4\",\"RRM1\",\"UNG\",\"GINS2\",\"MCM6\",\"CDCA7\",\"DTL\",\"PRIM1\",\"UHRF1\",\"HELLS\",\"RFC2\",\"RPA2\",\"NASP\",\"RAD51AP1\",\"GMNN\",\"WDR76\",\"SLBP\",\"CCNE2\",\"UBR7\",\"POLD3\",\"MSH2\",\"ATAD2\",\"RAD51\",\"RRM2\",\"CDC45\",\"CDC6\",\"EXO1\",\"TIPIN\",\"DSCC1\",\"BLM\",\"CASP8AP2\",\"USP1\",\"CLSPN\",\"POLA1\",\"CHAF1B\",\"BRIP1\",\"E2F8\"]\n",
      "    if g2m_genes is None: #use default list\n",
      "        g2m_genes =[\"HMGB2\",\"CDK1\",\"NUSAP1\",\"UBE2C\",\"BIRC5\",\"TPX2\",\"TOP2A\",\"NDC80\",\"CKS2\",\"NUF2\",\"CKS1B\",\"MKI67\",\"TMPO\",\"CENPF\",\"TACC3\",\"FAM64A\",\"SMC4\",\"CCNB2\",\"CKAP2L\",\"CKAP2\",\"AURKB\",\"BUB1\",\"KIF11\",\"ANP32E\",\"TUBB4B\",\"GTSE1\",\"KIF20B\",\"HJURP\",\"CDCA3\",\"HN1\",\"CDC20\",\"TTK\",\"CDC25C\",\"KIF2C\",\"RANGAP1\",\"NCAPD2\",\"DLGAP5\",\"CDCA2\",\"CDCA8\",\"ECT2\",\"KIF23\",\"HMMR\",\"AURKA\",\"PSRC1\",\"ANLN\",\"LBR\",\"CKAP5\",\"CENPE\",\"CTCF\",\"NEK2\",\"G2E3\",\"GAS2L3\",\"CBX5\",\"CENPA\"]\n",
      "    s_genes =  [geneprefix  + x for x in s_genes]\n",
      "    g2m_genes =  [geneprefix  + x for x in g2m_genes]\n",
      "    sc.tl.score_genes_cell_cycle(adata, s_genes=s_genes, g2m_genes=g2m_genes, use_raw = use_raw_data)\n",
      "    return(adata)\n",
      "\n",
      "\u001b[31;43;1mExecution:\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# save object\n",
    "adata = lpy.addCycleCycleAnnotation(adata, doinspect=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[\"phase_colors\"] = ['#6E40AA', '#FF8C38', '#28EA8D']\n",
    "adata.write_h5ad(\"N1-integrated_organoids.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
